From 9f78c821edf12c8c91c290f591cd5a0681853592 Mon Sep 17 00:00:00 2001
From: feng <feng@ingrasys>
Date: Wed, 16 Aug 2017 19:49:48 +0800
Subject: [PATCH] driver pca954x i2c mux force deselect on exit flag

---
 drivers/i2c/muxes/pca954x.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/i2c/muxes/pca954x.c b/drivers/i2c/muxes/pca954x.c
index c9db544..67b7020 100644
--- a/drivers/i2c/muxes/pca954x.c
+++ b/drivers/i2c/muxes/pca954x.c
@@ -240,7 +240,7 @@ static int pca954x_probe(struct i2c_client *client,
 		data->virt_adaps[num] =
 			i2c_add_mux_adapter(adap, &client->dev, client,
 				force, num, pca954x_select_chan,
-				(pdata && pdata->modes[num].deselect_on_exit) || deselect_on_exit
+				(pdata && pdata->modes[num].deselect_on_exit) || deselect_on_exit || 1
 					    ? pca954x_deselect_mux : NULL);
 
 		if (data->virt_adaps[num] == NULL) {
-- 
2.1.4

