Linux backport patch. Includes following commits:
a4dffccb72a7fa46bb0d7f29e607375387e09956


diff -Nur a/drivers/hwmon/pmbus/pmbus.h b/drivers/hwmon/pmbus/pmbus.h
--- a/drivers/hwmon/pmbus/pmbus.h	2017-11-09 16:25:22.760993964 +0000
+++ b/drivers/hwmon/pmbus/pmbus.h	2017-11-09 16:26:02.568994492 +0000
@@ -341,7 +341,7 @@
 #define PMBUS_HAVE_STATUS_VMON	BIT(19)
 
 enum pmbus_data_format { linear = 0, direct, vid };
-enum vrm_version { vr11 = 0, vr12 };
+enum vrm_version { vr11 = 0, vr12, vr13 };
 
 struct pmbus_driver_info {
 	int pages;		/* Total number of pages */
diff -Nur a/drivers/hwmon/pmbus/pmbus_core.c b/drivers/hwmon/pmbus/pmbus_core.c
--- a/drivers/hwmon/pmbus/pmbus_core.c	2017-11-09 16:25:22.760993964 +0000
+++ b/drivers/hwmon/pmbus/pmbus_core.c	2017-11-09 16:26:02.568994492 +0000
@@ -531,6 +531,10 @@
 		if (val >= 0x01)
 			rv = 250 + (val - 1) * 5;
 		break;
+	case vr13:
+		if (val >= 0x01)
+			rv = 500 + (val - 1) * 10;
+		break;
 	}
 	return rv;
 }
