From 95439e6f4f2f5e752bee76685627c78329fc8728 Mon Sep 17 00:00:00 2001
From: Alex Doyle <adoyle@nvidia.com>
Date: Mon, 30 Aug 2021 22:53:12 +0000
Subject: [PATCH] Fix 'unsigned char' changes value from '496' to '240'

---
 src/efikeygen.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/efikeygen.c b/src/efikeygen.c
index 121a238..848480a 100644
--- a/src/efikeygen.c
+++ b/src/efikeygen.c
@@ -208,7 +208,7 @@ static int
 add_cert_type(cms_context *cms, void *extHandle, int is_ca)
 {
 	SECItem bitStringValue;
-	unsigned char type = NS_CERT_TYPE_APP;
+	int type = NS_CERT_TYPE_APP;
 
 	if (is_ca)
 		type |= NS_CERT_TYPE_SSL_CA |
-- 
2.30.2

