From 5e18c5e057d08c7aeffb41d4c35583094fe67365 Mon Sep 17 00:00:00 2001
From: Alex Doyle <adoyle@nvidia.com>
Date: Mon, 30 Aug 2021 22:59:24 +0000
Subject: [PATCH] pesigcheck: Fix a wrong assignment.

---
 src/pesigcheck.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/pesigcheck.c b/src/pesigcheck.c
index 524cce3..8fa0f1a 100644
--- a/src/pesigcheck.c
+++ b/src/pesigcheck.c
@@ -318,7 +318,7 @@ check_signature(pesigcheck_context *ctx, int *nreasons,
 			reason->type = SIGNATURE;
 			reason->sig.data = data;
 			reason->sig.len = datalen;
-			reason->type = siBuffer;
+			reason->sig.type = siBuffer;
 			nreason += 1;
 			is_invalid = true;
 		}
@@ -330,7 +330,7 @@ check_signature(pesigcheck_context *ctx, int *nreasons,
 			reason->type = SIGNATURE;
 			reason->sig.data = data;
 			reason->sig.len = datalen;
-			reason->type = siBuffer;
+			reason->sig.type = siBuffer;
 			nreason += 1;
 			has_valid_cert = true;
 		}
-- 
2.30.2

