--- qemu-2.4.0/hw/arm/vexpress.c	2015-08-11 12:11:06.000000000 -0700
+++ qemu-2.4.0-new/hw/arm/vexpress.c	2015-09-02 12:55:46.644329535 -0700
@@ -521,11 +521,11 @@
     }
 
     qdev_prop_set_uint32(dev, "num-blocks",
                          VEXPRESS_FLASH_SIZE / VEXPRESS_FLASH_SECT_SIZE);
     qdev_prop_set_uint64(dev, "sector-length", VEXPRESS_FLASH_SECT_SIZE);
-    qdev_prop_set_uint8(dev, "width", 4);
+    qdev_prop_set_uint8(dev, "width", 2);
     qdev_prop_set_uint8(dev, "device-width", 2);
     qdev_prop_set_bit(dev, "big-endian", false);
     qdev_prop_set_uint16(dev, "id0", 0x89);
     qdev_prop_set_uint16(dev, "id1", 0x18);
     qdev_prop_set_uint16(dev, "id2", 0x00);
