SYS_EEPROM configs for platform

diff --git a/miscutils/sys_eeprom_alt.c b/miscutils/sys_eeprom_alt.c
index df55d18..c758f1e 100644
--- a/miscutils/sys_eeprom_alt.c
+++ b/miscutils/sys_eeprom_alt.c
@@ -14,6 +14,7 @@ int ipmi_read_sys_eeprom(void *eeprom_data, int offset, int len);
 int ipmi_write_sys_eeprom(void *eeprom_data, int len);
 int is_bmc_available(void);
 
+#define OLD_SYS_EEPROM_SIZE (512)
 int ipmi_read_sys_eeprom(void *eeprom_data, int offset, int len)
 {
     int i;
@@ -32,7 +33,7 @@ int ipmi_read_sys_eeprom(void *eeprom_data, int offset, int len)
             ipmi_len = DATA_SIZE;
         }
         memset(cmd, 0, cmd_length);
-        if (SYS_EEPROM_MAX_SIZE > 256) {
+        if (OLD_SYS_EEPROM_SIZE > 256) {
             sprintf(cmd, "%s 0x%x 0x%x 0x%x", EEPROM_READ,
                     (ipmi_offset>>8)&0x0ff, ipmi_offset&0x0ff, ipmi_len);
         } else {
@@ -82,7 +83,7 @@ int ipmi_write_sys_eeprom(void *eeprom_data, int len)
             ipmi_len = DATA_SIZE;
         }
         memset(cmd, 0, cmd_length);
-        if (SYS_EEPROM_MAX_SIZE > 256) {
+        if (OLD_SYS_EEPROM_SIZE > 256) {
             sprintf(cmd, "%s 0x%x 0x%x 0x%x", EEPROM_WRITE,
                     (ipmi_offset>>8)&0x0ff, ipmi_offset&0x0ff, ipmi_len);
         } else {
