SYS_EEPROM configs for platform

diff --git a/include/onie_tlvinfo.h b/include/onie_tlvinfo.h
index 055cf32..349b27d 100644
--- a/include/onie_tlvinfo.h
+++ b/include/onie_tlvinfo.h
@@ -173,6 +173,6 @@ bool tlvinfo_add_tlv(u_int8_t * eeprom, int tcode, char * strval);
 bool tlvinfo_decode_tlv(u_int8_t *eeprom, u_int8_t tcode, char* value);
 void show_tlv_code_list(void);
 
-#define OLD_SYS_EEPROM_SIZE (256)
+#define OLD_SYS_EEPROM_SIZE (512)
 #define OLD_SYS_EEPROM_I2C_ADDR (0x56)
 #define QSFP_ROOT_MUX_ADDR (0x77)
