From 765cd444a79392a58e940ad64d79df06d8a5245c Mon Sep 17 00:00:00 2001
From: Will Chao <will.chao@wnc.com.tw>
Date: Thu, 23 Sep 2021 17:28:25 +0800
Subject: [PATCH] Enlarge CONFIG_SYS_BOOTM_LEN, add bootm len dbg message

---
 common/bootm.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/common/bootm.c b/common/bootm.c
index adb1213..79ad729 100644
--- a/common/bootm.c
+++ b/common/bootm.c
@@ -32,7 +32,7 @@
 
 #ifndef CONFIG_SYS_BOOTM_LEN
 /* use 8MByte as default max gunzip size */
-#define CONFIG_SYS_BOOTM_LEN	0x800000
+#define CONFIG_SYS_BOOTM_LEN	0x40000000
 #endif
 
 #define IH_INITRD_ARCH IH_ARCH_DEFAULT
@@ -320,7 +320,7 @@ static int handle_decomp_error(int comp_type, size_t uncomp_size,
 	const char *name = genimg_get_comp_name(comp_type);
 
 	if (uncomp_size >= unc_len)
-		printf("Image too large: increase CONFIG_SYS_BOOTM_LEN\n");
+		printf("Image too large: increase CONFIG_SYS_BOOTM_LEN (uncomp_size=%d, unc_len=%d)\n", uncomp_size, unc_len);
 	else
 		printf("%s: uncompress error %d\n", name, ret);
 
-- 
2.17.1

