Busybox patch for onie syseeprom tool

diff --git a/include/sys_eeprom.h b/include/sys_eeprom.h
index 94dfd5a6ff23..ca650e6ed950 100644
--- a/include/sys_eeprom.h
+++ b/include/sys_eeprom.h
@@ -1,2 +1,14 @@
 int read_sys_eeprom(void *eeprom_data, int offset, int len);
 int write_sys_eeprom(void *eeprom_data, int len);
+
+#define SYS_EEPROM_I2C_SW1_I2C_BUS	0
+#define SYS_EEPROM_I2C_SW1_I2C_ADDR	0x73
+#define SYS_EEPROM_I2C_SW1_CH_SEL_REG	0
+#define SYS_EEPROM_I2C_SW1_CH_SEL_VAL	0x40
+
+#define SYS_EEPROM_I2C_SW2_I2C_BUS	0
+#define SYS_EEPROM_I2C_SW2_I2C_ADDR	0x71
+#define SYS_EEPROM_I2C_SW2_CH_SEL_REG	0
+#define SYS_EEPROM_I2C_SW2_CH_SEL_VAL	0x01
+
+#define SYS_EEPROM_I2C_SW_CMD_SZ	50
diff --git a/miscutils/sys_eeprom.c b/miscutils/sys_eeprom.c
index e34abf59782e..09141e34d41a 100644
--- a/miscutils/sys_eeprom.c
+++ b/miscutils/sys_eeprom.c
@@ -1,5 +1,6 @@
 #include "libbb.h"
 #include "onie_tlvinfo.h"
+#include "sys_eeprom.h"
 #include <getopt.h>
 
 static u_int8_t eeprom[SYS_EEPROM_SIZE];
@@ -53,6 +54,7 @@ int onie_syseeprom_main(int argc, char **argv)
     char *value, *subopts, *tname;
     int index, c, i, option_index, tcode;
     char tlv_value[TLV_DECODE_VALUE_MAX_LEN];
+    char i2c_sw_sel_cmd[SYS_EEPROM_I2C_SW_CMD_SZ];
 
     const size_t tlv_code_count = sizeof(tlv_code_list) /
 	sizeof(tlv_code_list[0]);
@@ -68,6 +70,24 @@ int onie_syseeprom_main(int argc, char **argv)
 	{0,         0,                    0,      0},
     };
 
+    /*
+     * Set the I2C switch config so that ID-eeprom channel is selected.
+     */
+    memset(i2c_sw_sel_cmd, 0, SYS_EEPROM_I2C_SW_CMD_SZ);
+    snprintf(i2c_sw_sel_cmd, SYS_EEPROM_I2C_SW_CMD_SZ, "i2cset -y %d 0x%x %d %d ",
+	SYS_EEPROM_I2C_SW1_I2C_BUS, SYS_EEPROM_I2C_SW1_I2C_ADDR,
+	SYS_EEPROM_I2C_SW1_CH_SEL_REG, SYS_EEPROM_I2C_SW1_CH_SEL_VAL);
+
+    system(i2c_sw_sel_cmd);
+    usleep(5000);
+
+    memset(i2c_sw_sel_cmd, 0, SYS_EEPROM_I2C_SW_CMD_SZ);
+    snprintf(i2c_sw_sel_cmd, SYS_EEPROM_I2C_SW_CMD_SZ, "i2cset -y %d 0x%x %d %d ",
+	SYS_EEPROM_I2C_SW2_I2C_BUS, SYS_EEPROM_I2C_SW2_I2C_ADDR,
+	SYS_EEPROM_I2C_SW2_CH_SEL_REG, SYS_EEPROM_I2C_SW2_CH_SEL_VAL);
+
+    system(i2c_sw_sel_cmd);
+
     for (i = 0; i < tlv_code_count; i++) {
 	    tokens[i*2] = malloc(6);
 	    snprintf(tokens[(i*2)], 6,     "0x%x", tlv_code_list[i].m_code);
diff --git a/miscutils/sys_eeprom_i2c.c b/miscutils/sys_eeprom_i2c.c
index ed3235bcc1c8..46db9e3c41b3 100644
--- a/miscutils/sys_eeprom_i2c.c
+++ b/miscutils/sys_eeprom_i2c.c
@@ -55,6 +55,7 @@ int write_sys_eeprom(void *eeprom_data, int len)
 	    return -1;
 	}
 	eeprom_write_byte(&e, addr, *c);
+	usleep(5000);
 	eeprom_close(&e);
 	c++; addr++;
     }
