diff --git a/drivers/mmc/host/sdhci-of-esdhc.c b/drivers/mmc/host/sdhci-of-esdhc.c
index d6799a8..22dba50 100644
--- a/drivers/mmc/host/sdhci-of-esdhc.c
+++ b/drivers/mmc/host/sdhci-of-esdhc.c
@@ -739,6 +739,9 @@ static int sdhci_esdhc_probe(struct platform_device *pdev)
 		esdhc_clock_control(host, true);
 	}
 
+	host->mmc->caps |= MMC_CAP_NONREMOVABLE;
+	host->quirks |= SDHCI_QUIRK_BROKEN_CARD_DETECTION;
+
 	ret = sdhci_add_host(host);
 	if (ret)
 		sdhci_pltfm_free(pdev);
