

diff --git a/drivers/acpi/osl.c b/drivers/acpi/osl.c
index f31c5c5..569a835 100644
--- a/drivers/acpi/osl.c
+++ b/drivers/acpi/osl.c
@@ -1255,7 +1255,7 @@ __setup("acpi_serialize", acpi_serialize_setup);
 #define ENFORCE_RESOURCES_LAX    1
 #define ENFORCE_RESOURCES_NO     0
 
-static unsigned int acpi_enforce_resources = ENFORCE_RESOURCES_STRICT;
+static unsigned int acpi_enforce_resources = ENFORCE_RESOURCES_LAX;
 
 static int __init acpi_enforce_resources_setup(char *str)
 {
