From 9839e08de25adcf2e6ba79640b47daa3eb03e2cb Mon Sep 17 00:00:00 2001
From: Varun Sethi <Varun.Sethi@freescale.com>
Date: Sun, 23 Aug 2015 19:23:32 +0530
Subject: [PATCH 174/245] Fix copyright for pci-layerscape.h.

Signed-off-by: Varun Sethi <Varun.Sethi@freescale.com>
---
 drivers/pci/host/pci-layerscape.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/pci/host/pci-layerscape.h b/drivers/pci/host/pci-layerscape.h
index 8c0642a..7ce1ef9 100644
--- a/drivers/pci/host/pci-layerscape.h
+++ b/drivers/pci/host/pci-layerscape.h
@@ -1,4 +1,8 @@
 /*
+ * Copyright (C) 2015 Freescale Semiconductor.
+ *
+ * Author: Varun Sethi <Varun.Sethi@freescale.com>
+ *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License version 2 as
  * published by the Free Software Foundation.
-- 
2.1.0.27.g96db324

