From 574a9cee81233b5fe0c12a3233cbb712c575753e Mon Sep 17 00:00:00 2001
From: Mingkai Hu <Mingkai.Hu@freescale.com>
Date: Thu, 19 Nov 2015 12:49:50 +0800
Subject: [PATCH 187/245] arm64/layerscape: Enable PCIe for Layerscape

Signed-off-by: Mingkai Hu <Mingkai.Hu@freescale.com>
---
 arch/arm64/Kconfig | 1 +
 1 file changed, 1 insertion(+)

diff --git a/arch/arm64/Kconfig b/arch/arm64/Kconfig
index 38b52a0..c5ad326 100644
--- a/arch/arm64/Kconfig
+++ b/arch/arm64/Kconfig
@@ -182,6 +182,7 @@ config ARCH_EXYNOS7
 
 config ARCH_LAYERSCAPE
 	bool "Freescale Layerscape SoC family"
+	select PCI_LAYERSCAPE if PCI
 	help
 	  This enables support for Freescale Layerscape SoC family
 
-- 
2.1.0.27.g96db324

