Support IX7/IX8 motherboard detect

diff --git a/include/mb_detect.h b/include/mb_detect.h
index ba02829..ad9a198 100644
--- a/include/mb_detect.h
+++ b/include/mb_detect.h
@@ -8,10 +8,12 @@ typedef __s32 s32;
 #define MB_DETECT_I2C_BUS 0
 
 #define MAX_NUM_MB 0x3f
-#define CPU_BOARD_CAPABILITY "ix1b,ly7"
+#define CPU_BOARD_CAPABILITY "ix1b,ly7,ix7,ix8"
 enum {
 	MB_IX1B = 0,
 	MB_LY7,
+	MB_IX7,
+	MB_IX8,
 	MB_LAST
 };
 
diff --git a/miscutils/mb_detect.c b/miscutils/mb_detect.c
index f8e07b3..fab78fa 100644
--- a/miscutils/mb_detect.c
+++ b/miscutils/mb_detect.c
@@ -13,6 +13,8 @@ MBDetectPath mbDetectPath[] = {
 SysBoardInfo sysBoardInfo[MAX_NUM_MB] = {
 	{ 0x18,  "ix1b_rglbmc",  "IX1B\0",   "IX1B-RGLBMC\0",    0x0000c1ba },
 	{ 0x11,  "ly7_rglbmc",  "LY7\0",   "LY7-RGLBMC\0",    0x0000c1ba },
+	{ 0x1C,  "ix7_rglbmc",  "IX7\0",   "IX7-RGLBMC\0",    0x0000c70a },
+	{ 0x1D,  "ix8_rglbmc",  "IX8\0",   "IX8-RGLBMC\0",    0x0000c80a },
 };
 
 SysEEPROMPath sysEEPROMPath[] = {

