diff --git a/drivers/acpi/pci_link.c b/drivers/acpi/pci_link.c
old mode 100644
new mode 100755
index bc3d914d..60d27e2f
--- a/drivers/acpi/pci_link.c
+++ b/drivers/acpi/pci_link.c
@@ -173,7 +173,7 @@ static int acpi_pci_link_get_possible(struct acpi_pci_link *link)
 	status = acpi_walk_resources(link->device->handle, METHOD_NAME__PRS,
 				     acpi_pci_link_check_possible, link);
 	if (ACPI_FAILURE(status)) {
-		ACPI_EXCEPTION((AE_INFO, status, "Evaluating _PRS"));
+		acpi_handle_debug(link->device->handle, "failed to evaluate _PRS");
 		return -ENODEV;
 	}
 
