Add support for quanta_bwde

diff --git a/include/mb_detect.h b/include/mb_detect.h
index a223396..c4b8a90 100644
--- a/include/mb_detect.h
+++ b/include/mb_detect.h
@@ -8,17 +8,11 @@ typedef __s32 s32;
 #define MB_DETECT_I2C_BUS 0
 
 #define MAX_NUM_MB 0x3f
-#define CPU_BOARD_CAPABILITY "ly6,ly8,ly5a,ly2a,ly9,ly9a,ly6f,ix1,ix2"
+#define CPU_BOARD_CAPABILITY "ix3,ix4,ix9"
 enum {
-	MB_LY6 = 0,
-	MB_LY8,
-	MB_LY5A,
-	MB_LY2A,
-	MB_LY9,
-	MB_LY9A,
-	MB_LY6F,
-	MB_IX1,
-	MB_IX2,
+	MB_IX3 = 0,
+	MB_IX4,
+	MB_IX9,
 	MB_LAST
 };
 
diff --git a/miscutils/mb_detect.c b/miscutils/mb_detect.c
index d47489e..ea124a1 100644
--- a/miscutils/mb_detect.c
+++ b/miscutils/mb_detect.c
@@ -7,27 +7,18 @@ extern void mbdetect_usage(void);
 
 MBDetectPath mbDetectPath[] = {
 	{ 0x72, 2, 0x23, I2C_DEVICE_PCA9555_LENGTH, 1, 2 },
-	{ 0x77, 0, 0x23, I2C_DEVICE_PCA9555_LENGTH, 1, 2 },
-	{ 0x77, 7, 0x23, I2C_DEVICE_PCA9555_LENGTH, 1, 2 },
-	{ 0x77, 2, 0x23, I2C_DEVICE_PCA9555_LENGTH, 1, 2 },
+	{ 0x72, 3, 0x23, I2C_DEVICE_PCA9555_LENGTH, 1, 2 },
+	{ 0x72, 3, 0x74, I2C_DEVICE_PCA9555_LENGTH, 1, 2 },
 };
 
 SysBoardInfo sysBoardInfo[MAX_NUM_MB] = {
-	{ 0x00, "ly6_rangeley", "LY6\0", "LY6-RANGELEY\0", 0x0000f605 },
-	{ 0x01, "ly8_rangeley", "LY8\0", "LY8-RANGELEY\0", 0x0000f805 },
-	{ 0x09, "ly5a_rangeley", "LY5A\0", "LY5A-RANGELEY\0", 0x0000f5a5 },
-	{ 0x08, "ly2a_rangeley", "LY2A\0", "LY2A-RANGELEY\0", 0x0000f2a5 },
-	{ 0x02, "ly9_rangeley",  "LY9\0",  "LY9-RANGELEY\0",  0x0000f905 },
-	{ 0x0e, "ly9a_rangeley",  "LY9A\0",  "LY9A-RANGELEY\0",  0x0000f9a5 },
-	{ 0x05, "ly6f_rangeley",  "LY6F\0",  "LY6F-RANGELEY\0",  0x0000f6f5 },
-	{ 0x0d, "ix1_rangeley",  "IX1\0",  "IX1-RANGELEY\0",   0x0000c105 },
-	{ 0x0f, "ix2_rangeley",  "IX2\0",  "IX2-RANGELEY\0",  0x0000c205 },
+	{ 0x1f, "ix3_bwde",  "IX3\0",  "IX3-BWDE\0",  0x0000c308 },
+	{ 0x19, "ix4_bwde",  "IX4\0",  "IX4-BWDE\0",  0x0000c408 },
+	{ 0x20, "ix9_bwde",  "IX9\0",  "IX9-BWDE\0",  0x0000c908 },
 };
 
 SysEEPROMPath sysEEPROMPath[] = {
 	{ 0x72, 2, 0x54},
-	{ 0x72, 3, 0x54},
-	{ 0x77, 2, 0x54},
 };
 
 /*
@@ -261,14 +250,7 @@ s32 sys_eeprom_path_get(SysEEPROMPath *ep) {
 
 	if(index < MB_LAST) {
 		switch(index) {
-			case MB_LY6:
-			case MB_LY8:
-			case MB_LY2A:
-			case MB_LY9:
-			case MB_LY9A:
-			case MB_LY6F:
-				memcpy(ep, &sysEEPROMPath[1], sizeof(SysEEPROMPath));
-				break;
+			/* TBD : when boardinfo EEPROM path different, should define a case */
 
 			default:
 				memcpy(ep, &sysEEPROMPath[0], sizeof(SysEEPROMPath));

