From e2033e9ab973b07f3048abc7549b580832422913 Mon Sep 17 00:00:00 2001
From: Zhao Qiang <qiang.zhao@nxp.com>
Date: Wed, 13 Dec 2017 17:36:24 +0800
Subject: [PATCH 057/129] phy: atheros: set auto-negotiation for AR8021

Signed-off-by: Zhao Qiang <qiang.zhao@nxp.com>
---
 drivers/net/phy/atheros.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/drivers/net/phy/atheros.c b/drivers/net/phy/atheros.c
index b34cdd3..d7e76de 100644
--- a/drivers/net/phy/atheros.c
+++ b/drivers/net/phy/atheros.c
@@ -19,6 +19,7 @@
 
 static int ar8021_config(struct phy_device *phydev)
 {
+	phy_write(phydev, MDIO_DEVAD_NONE, 0x00, 0x1200);
 	phy_write(phydev, MDIO_DEVAD_NONE, 0x1d, 0x05);
 	phy_write(phydev, MDIO_DEVAD_NONE, 0x1e, 0x3D47);
 
-- 
1.7.1

