From a6932b4eeef41de15c92bf581235cfeb20bd42cc Mon Sep 17 00:00:00 2001
From: Pankit Garg <pankit.garg@nxp.com>
Date: Tue, 16 Jan 2018 11:56:02 +0530
Subject: [PATCH 104/129] armv8: ls1088a: vid: Add some function in SPL for VID

Signed-off-by: Pankit Garg <pankit.garg@nxp.com>
---
 board/freescale/ls1088a/ls1088a.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/board/freescale/ls1088a/ls1088a.c b/board/freescale/ls1088a/ls1088a.c
index 36e5bf1..88c07d1 100644
--- a/board/freescale/ls1088a/ls1088a.c
+++ b/board/freescale/ls1088a/ls1088a.c
@@ -211,6 +211,7 @@ unsigned long get_board_ddr_clk(void)
 
 	return 66666666;
 }
+#endif
 
 int select_i2c_ch_pca9547(u8 ch)
 {
@@ -225,6 +226,7 @@ int select_i2c_ch_pca9547(u8 ch)
 	return 0;
 }
 
+#if !defined(CONFIG_SPL_BUILD)
 void board_retimer_init(void)
 {
 	u8 reg;
@@ -326,6 +328,7 @@ int misc_init_r(void)
 	return 0;
 }
 #endif
+#endif
 
 int i2c_multiplexer_select_vid_channel(u8 channel)
 {
@@ -449,6 +452,7 @@ int init_func_vid(void)
 }
 #endif
 
+#if !defined(CONFIG_SPL_BUILD)
 int board_init(void)
 {
 	init_final_memctl_regs();
-- 
1.7.1

