From 1e96fd8f464dfe23eb692a11018c20d70546783b Mon Sep 17 00:00:00 2001
From: Priyanka Jain <priyanka.jain@nxp.com>
Date: Tue, 18 Sep 2018 10:49:36 +0530
Subject: [PATCH 107/107] ls2080ardb: remove dhcp function from u-boot env as boot source

Signed-off-by: Priyanka Jain <priyanka.jain@nxp.com>
---
 include/configs/ls2080ardb.h |    3 +--
 1 files changed, 1 insertions(+), 2 deletions(-)

diff --git a/include/configs/ls2080ardb.h b/include/configs/ls2080ardb.h
index 299362c..c8bb2ef 100644
--- a/include/configs/ls2080ardb.h
+++ b/include/configs/ls2080ardb.h
@@ -334,8 +334,7 @@ unsigned long get_board_sys_clk(void);
 #define BOOT_TARGET_DEVICES(func) \
 	func(USB, usb, 0) \
 	func(MMC, mmc, 0) \
-	func(SCSI, scsi, 0) \
-	func(DHCP, dhcp, na)
+	func(SCSI, scsi, 0)
 #include <config_distro_bootcmd.h>
 
 #ifdef CONFIG_QSPI_BOOT
-- 
1.7.1

