From 6a240aef1664214d40a15a45088b0ca06e76f4c2 Mon Sep 17 00:00:00 2001
From: Shengzhou Liu <Shengzhou.Liu@nxp.com>
Date: Mon, 7 Aug 2017 15:03:52 +0800
Subject: [PATCH 127/211] arm64: ls1043ardb: Add sd_bootcmd for distro fallback in case of sdboot

Signed-off-by: Shengzhou Liu <Shengzhou.Liu@nxp.com>
---
 include/configs/ls1043a_common.h |   12 +++++++++++-
 1 files changed, 11 insertions(+), 1 deletions(-)

diff --git a/include/configs/ls1043a_common.h b/include/configs/ls1043a_common.h
index 030ebb0..dbb700b 100644
--- a/include/configs/ls1043a_common.h
+++ b/include/configs/ls1043a_common.h
@@ -284,6 +284,8 @@
 	"load_addr=0xa0000000\0"		\
 	"kernelheader_size=0x40000\0"		\
 	"kernel_size=0x2800000\0"		\
+	"kernel_addr_sd=0x8000\0"		\
+	"kernel_size_sd=0x14000\0"		\
 	"console=ttyS0,115200\0"		\
 	"mtdparts=" MTDPARTS_DEFAULT "\0"	\
 	BOOTENV					\
@@ -328,12 +330,20 @@
 		"$kernel_size; env exists secureboot "	\
 		"&& cp.b $kernelheader_addr $kernelheader_addr_r "	\
 		"$kernelheader_size && esbc_validate ${kernelheader_addr_r}; " \
-		"bootm $load_addr#$board\0"
+		"bootm $load_addr#$board\0"	    \
+	"sd_bootcmd=echo Trying load from SD ..;"   \
+		"mmcinfo; mmc read $load_addr "	    \
+		"$kernel_addr_sd $kernel_size_sd ;" \
+		" bootm $load_addr#$board\0"
+
 
 #undef CONFIG_BOOTCOMMAND
 #if defined(CONFIG_QSPI_BOOT) || defined(CONFIG_SD_BOOT_QSPI)
 #define CONFIG_BOOTCOMMAND "run distro_bootcmd; run qspi_bootcmd; "	\
 			   "env exists secureboot && esbc_halt;"
+#elif defined(CONFIG_SD_BOOT)
+#define CONFIG_BOOTCOMMAND "run distro_bootcmd; run sd_bootcmd; "	\
+			   "env exists secureboot && esbc_halt;"
 #else
 #define CONFIG_BOOTCOMMAND "run distro_bootcmd; run nor_bootcmd; "	\
 			   "env exists secureboot && esbc_halt;"
-- 
1.7.1

