From 2dcba27d34d54dcd8dd60c5c4802979030b02c07 Mon Sep 17 00:00:00 2001
From: Vinitha Pillai-B57223 <vinitha.pillai@nxp.com>
Date: Wed, 27 Sep 2017 10:51:40 +0530
Subject: [PATCH 157/211] arm64: ls2088ardb: Fix nor fall back option failure

Signed-off-by: Vinitha Pillai-B57223 <vinitha.pillai@nxp.com>
---
 include/configs/ls2080ardb.h |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/include/configs/ls2080ardb.h b/include/configs/ls2080ardb.h
index 8a34990..41ec32f 100644
--- a/include/configs/ls2080ardb.h
+++ b/include/configs/ls2080ardb.h
@@ -443,7 +443,8 @@ unsigned long get_board_sys_clk(void);
 			"&& esbc_validate 0x20780000; "			\
 			"fsl_mc start mc 0x20a00000 0x20e00000 &&"	\
 			" fsl_mc lazyapply dpl 0x20d00000;"		\
-			"run distro_bootcmd;run qspi_bootcmd"
+			"run distro_bootcmd;run qspi_bootcmd;"		\
+			"env exists secureboot && esbc_halt;"
 #elif defined(CONFIG_SD_BOOT)
 #define CONFIG_BOOTCOMMAND						\
 			"env exists mcinitcmd && run mcinitcmd "	\
@@ -459,7 +460,7 @@ unsigned long get_board_sys_clk(void);
 			"&& esbc_validate 0x580780000; "		\
 			"fsl_mc start mc 0x580a00000 0x580e00000 &&"	\
 			" fsl_mc lazyapply dpl 0x580d00000;"		\
-			"run distro_bootcmd;run nor_bootcmd"		\
+			"run distro_bootcmd;run nor_bootcmd;"		\
 			"env exists secureboot && esbc_halt;"
 #endif
 
-- 
1.7.1

