Platform support for the AGEMA AG7648C

diff --git a/drivers/platform/x86/Kconfig b/drivers/platform/x86/Kconfig
index 7f43cf8..b1acf68 100644
--- a/drivers/platform/x86/Kconfig
+++ b/drivers/platform/x86/Kconfig
@@ -779,4 +779,17 @@ config SAMSUNG_Q10
 	  This driver provides support for backlight control on Samsung Q10
 	  and related laptops, including Dell Latitude X200.
 
+config AGEMA_AG7648C
+	tristate "AGEMA AG7648C  Platform Support"
+	depends on X86 && PCI
+	select I2C_ISCH
+	select I2C_ISMT
+	select GPIO_SCH
+	select MFD_CORE
+	select LPC_SCH
+	---help---
+	 Enable support for AGEMA AG7648C switching platform.  This
+	 platform has a C2338 Atom (Rangeley) CPU module, a Broadcom
+	 Trident2/Trident2+ switch ASIC. It has 48x10G Copper and 6x40G QSFP ports.
+
 endif # X86_PLATFORM_DEVICES
