diff --git a/machine/kvm_x86_64/machine.make b/machine/kvm_x86_64/machine.make
index 438d762e..804cf342 100644
--- a/machine/kvm_x86_64/machine.make
+++ b/machine/kvm_x86_64/machine.make
@@ -64,15 +64,15 @@ SKIP_ETHMGMT_MACS = yes
 #  that this provides.
 # The kernel/config-insecure file is provided as
 #  an example.
-SECURE_BOOT_ENABLE = yes
+SECURE_BOOT_ENABLE = no
 
 # Enable extended secure boot:
 #  Activates - ONIE password
-SECURE_BOOT_EXT = yes
+SECURE_BOOT_EXT = no
 
 # Enable GRUB verification of files and passwords
 # Requires secure boot
-SECURE_GRUB = yes
+SECURE_GRUB = no
 
 # Define the makefile with security settings, to
 # provide the option of using another file with different settings.
