--- a/sys_eeprom/sys_eeprom_platform.h	2016-11-24 17:40:04.969356863 +0200
+++ b/sys_eeprom/sys_eeprom_platform.h	2016-11-24 17:46:35.169364938 +0200
@@ -1,4 +1,10 @@
 #ifndef __SYS_EEPROM_PLATFORM_H_
 #define __SYS_EEPROM_PLATFORM_H_
 
+#define SYS_EEPROM_MAX_SIZE 4096
+#define SYS_EEPROM_OFFSET 0
+#define SYS_EEPROM_SIZE 1024
+#define SYS_EEPROM_I2C_DEVICE "/dev/i2c-8"
+#define SYS_EEPROM_I2C_ADDR 0x51
+
 #endif /* __SYS_EEPROM_PLATFORM_H_ */
--- a/sys_eeprom/onie_tlvinfo.c	2016-11-24 17:39:44.313356435 +0200
+++ b/sys_eeprom/onie_tlvinfo.c	2016-11-24 17:44:35.425362460 +0200
@@ -733,7 +733,7 @@
 
     c = eeprom_data;
     if (eeprom_open(SYS_EEPROM_I2C_DEVICE, SYS_EEPROM_I2C_ADDR,
-		    EEPROM_TYPE_8BIT_ADDR, &e)) {
+		    EEPROM_TYPE_16BIT_ADDR, &e)) {
 	printf("ERROR: Cannot open I2C device\n");
 	return -1;
     }
@@ -759,7 +759,7 @@
     c = eeprom_data;
     for (i = 0; i < len; i++) {
 	if (eeprom_open(SYS_EEPROM_I2C_DEVICE, SYS_EEPROM_I2C_ADDR,
-			EEPROM_TYPE_8BIT_ADDR, &e)) {
+			EEPROM_TYPE_16BIT_ADDR, &e)) {
 	    printf("ERROR: Cannot open I2C device\n");
 	    return -1;
 	}
--- a/eepromer/eeprog.c	2016-11-24 17:40:37.813357542 +0200
+++ b/eepromer/eeprog.c	2016-11-24 17:48:13.997366984 +0200
@@ -211,7 +211,7 @@
 		}
 	}
 	if(!eeprom_type)
-		eeprom_type = EEPROM_TYPE_8BIT_ADDR; // default
+		eeprom_type = EEPROM_TYPE_16BIT_ADDR; // default
 
 	usage_if(op == 0); // no switches 
 	// set device and i2c_addr reading from cmdline or env
