From 770c7d7b1433cd63238aa4b319c958b2c2b75fa5 Mon Sep 17 00:00:00 2001
From: Doron Tsur <doront@mellanox.com>
Date: Thu, 31 Jul 2014 13:46:56 +0300
Subject: [PATCH 2/3] fixing DEBUG compilation

Signed-off-by: Doron Tsur <doront@mellanox.com>
---
 sys_eeprom/onie_tlvinfo.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/sys_eeprom/onie_tlvinfo.c b/sys_eeprom/onie_tlvinfo.c
index cf88f70..ea5b30b 100644
--- a/sys_eeprom/onie_tlvinfo.c
+++ b/sys_eeprom/onie_tlvinfo.c
@@ -414,6 +414,9 @@ void show_eeprom(u_int8_t *eeprom)
     int curr_tlv;
     tlvinfo_header_t * eeprom_hdr = (tlvinfo_header_t *) eeprom;
     tlvinfo_tlv_t    * eeprom_tlv;
+#ifdef DEBUG
+    int i;
+#endif
 
     if ( !is_valid_tlvinfo_header(eeprom_hdr) ) {
 	printf("EEPROM does not contain data in a valid TlvInfo format.\n");
-- 
1.8.1.2

