diff --git a/sys_eeprom/sys_eeprom.c b/sys_eeprom/sys_eeprom.c
index eec083a..9f22a5b 100644
--- a/sys_eeprom/sys_eeprom.c
+++ b/sys_eeprom/sys_eeprom.c
@@ -83,7 +83,7 @@ int main(int argc, char * const argv[])
 
     for (i = 0; i < tlv_code_count; i++) {
 	    tokens[i] = (char *) malloc(6);
-	    sprintf(tokens[i], "0x%x", tlv_code_list[i].m_code);
+	    sprintf(tokens[i], "0x%X", tlv_code_list[i].m_code);
     }
     tokens[tlv_code_count] = NULL;
 
@@ -123,13 +123,13 @@ int main(int argc, char * const argv[])
 		for (i = 0; i < sizeof(tlv_code_default)/sizeof(struct tlv_code_desc); i++) {
 		//printf("Enter process i=%d\n",i);
 		    if (tlvinfo_delete_tlv(eeprom, tlv_code_default[i].m_code) == TRUE) {
-			    printf("Deleting TLV 0x%x: %s\n", tlv_code_default[i].m_code, tlv_code_list[i].m_name);
+			    printf("Deleting TLV 0x%X: %s\n", tlv_code_default[i].m_code, tlv_code_list[i].m_name);
 		    }
 			if (!tlvinfo_add_tlv(eeprom, tlv_code_default[i].m_code, tlv_code_default[i].m_name)) {
                             err = 1;
 				    goto syseeprom_err;
 			} else {
-			    printf("Adding   TLV 0x%x: %s\n", tlv_code_default[i].m_code, tlv_code_default[i].m_name);
+			    printf("Adding   TLV 0x%X: %s\n", tlv_code_default[i].m_code, tlv_code_default[i].m_name);
 			}
 			//printf("i=%d\n",i);
 		}		
@@ -152,14 +152,14 @@ int main(int argc, char * const argv[])
 			}
 		    }
 		    if (tlvinfo_delete_tlv(eeprom, tcode) == TRUE) {
-			    printf("Deleting TLV 0x%x: %s\n", tcode, tname);
+			    printf("Deleting TLV 0x%X: %s\n", tcode, tname);
 		    }
 		    if (value) {
 			if (!tlvinfo_add_tlv(eeprom, tcode, value)) {
                             err = 1;
 			    goto syseeprom_err;
 			} else {
-			    printf("Adding   TLV 0x%x: %s\n", tcode, tname);
+			    printf("Adding   TLV 0x%X: %s\n", tcode, tname);
 			}
 		    }
 		    update = 1;
