From 6f439d8a5156691c0eaf64d0337ebb5b514b79c3 Mon Sep 17 00:00:00 2001
Subject: [PATCH 03/16] Fix wrong indent typo.

Signed-off-by: Peter Jones <pjones@redhat.com>
Signed-off-by: Curt Brune <curt@cumulusnetworks.com>

diff --git a/src/dp-media.c b/src/dp-media.c
index 07e09ab..a677204 100644
--- a/src/dp-media.c
+++ b/src/dp-media.c
@@ -111,7 +111,7 @@ _format_media_dn(char *buf, size_t size, const_efidp dp)
 		for (int i = 0; !efi_guid_is_zero(&subtypes[i].guid); i++) {
 			if (efi_guid_cmp(&subtypes[i].guid,
 					  &dp->ramdisk.disk_type_guid))
-			continue;
+				continue;
 
 			label = subtypes[i].label;
 			break;
