From 6b0d73d87b49d12724cb131e1fe8327636a1f6db Mon Sep 17 00:00:00 2001
Subject: [PATCH 05/16] Minor formatting change in efiboot-loadopt.h

Signed-off-by: Peter Jones <pjones@redhat.com>
Signed-off-by: Curt Brune <curt@cumulusnetworks.com>

diff --git a/src/include/efivar/efiboot-loadopt.h b/src/include/efivar/efiboot-loadopt.h
index 1a67773..93d9fe6 100644
--- a/src/include/efivar/efiboot-loadopt.h
+++ b/src/include/efivar/efiboot-loadopt.h
@@ -51,16 +51,16 @@ extern int efi_load_option_optional_data(efi_load_option *opt, size_t opt_size,
 	__attribute__((__nonnull__ (1,3)))
 	__attribute__((__visibility__ ("default")));
 
-extern ssize_t
-efi_load_option_args_from_file(uint8_t *buf, ssize_t size, char *filename)
+extern ssize_t efi_load_option_args_from_file(uint8_t *buf, ssize_t size,
+					      char *filename)
 	__attribute__((__nonnull__ (3)))
 	__attribute__((__visibility__ ("default")));
-extern ssize_t
-efi_load_option_args_as_utf8(uint8_t *buf, ssize_t size, char *utf8)
+extern ssize_t efi_load_option_args_as_utf8(uint8_t *buf, ssize_t size,
+					    char *utf8)
 	__attribute__((__nonnull__ (3)))
 	__attribute__((__visibility__ ("default")));
-extern ssize_t
-efi_load_option_args_as_ucs2(uint16_t *buf, ssize_t size, char *utf8)
+extern ssize_t efi_load_option_args_as_ucs2(uint16_t *buf, ssize_t size,
+					    char *utf8)
 	__attribute__((__nonnull__ (3)))
 	__attribute__((__visibility__ ("default")));
 
