From 51b94a976d73c72a958e4450723bba219f60b972 Mon Sep 17 00:00:00 2001
Subject: [PATCH 06/16] Missed one tiny bit of getting rid of -lpci.

Signed-off-by: Peter Jones <pjones@redhat.com>
Signed-off-by: Curt Brune <curt@cumulusnetworks.com>

diff --git a/src/creator.c b/src/creator.c
index 4d51f4f..2aea5be 100644
--- a/src/creator.c
+++ b/src/creator.c
@@ -23,7 +23,6 @@
 #include <mntent.h>
 #include <netinet/in.h>
 #include <netinet/ip.h>
-#include <pci/pci.h>
 #include <stdlib.h>
 #include <stdio.h>
 #include <sys/types.h>
