From 95692078777e31d9667a5d87d1cb88849247367f Mon Sep 17 00:00:00 2001
Subject: [PATCH 12/16] Get rid of extra _dp_ from efidp_is_multiinstance()

Signed-off-by: Peter Jones <pjones@redhat.com>
Signed-off-by: Curt Brune <curt@cumulusnetworks.com>

diff --git a/src/include/efivar/efivar-dp.h b/src/include/efivar/efivar-dp.h
index 9fccdb2..312d404 100644
--- a/src/include/efivar/efivar-dp.h
+++ b/src/include/efivar/efivar-dp.h
@@ -761,7 +761,7 @@ efidp_next_instance(const_efidp in, const_efidp *out)
 
 static inline int
 __attribute__((__unused__))
-efidp_dp_is_multiinstance(const_efidp dn)
+efidp_is_multiinstance(const_efidp dn)
 {
 	while (1) {
 		const_efidp next;
