Modify getCpuId() method

diff --git a/sys_eeprom/i2c_chips.c b/sys_eeprom/i2c_chips.c
index b39228b..cb07016 100755
--- a/sys_eeprom/i2c_chips.c
+++ b/sys_eeprom/i2c_chips.c
@@ -121,16 +121,10 @@ unsigned char getCpuId()
 	else
 	{
 		gpio_init(19);
-		gpio_init(20);
-		gpio_init(21);
 		gpio_set_dir(19,gpio_out);
-		gpio_set_dir(20,gpio_out);
-		gpio_set_dir(21,gpio_out);
 	}
 	/*read cpu id*/
 	cpu_id |= gpio_get_value(19) << 0 ;
-	cpu_id |= gpio_get_value(20) << 1 ;
-	cpu_id |= gpio_get_value(21) << 2 ;
 	if (iopl(0))
 	{		
 		perror("iopl");
