From b25c0ab4cc8c321cddde9fb38048a23c571cdda3 Mon Sep 17 00:00:00 2001
Subject: [PATCH 03/12] Fix MAC() device path printing.

67491a308 accidentally made it print MAC(MAC( .

Signed-off-by: Peter Jones <pjones@redhat.com>
Signed-off-by: Curt Brune <curt@cumulusnetworks.com>

diff --git a/src/lib/unparse_path.c b/src/lib/unparse_path.c
index a8eb0c1..d76191d 100644
--- a/src/lib/unparse_path.c
+++ b/src/lib/unparse_path.c
@@ -233,13 +233,6 @@ unparse_messaging_path(char *buffer, size_t buffer_size, EFI_DEVICE_PATH *path)
 			return needed;
 		buf_offset += needed;
 
-		needed = snprintf(buffer + buf_offset,
-			buffer_size == 0 ? 0 : buffer_size - buf_offset,
-			"MAC(");
-		if (needed < 0)
-			return needed;
-		buf_offset += needed;
-
 		needed = unparse_raw(buffer + buf_offset,
 			buffer_size == 0 ? 0 : buffer_size - buf_offset,
 			mac->macaddr, 6);
