diff --git a/drivers/acpi/osl.c b/drivers/acpi/osl.c
index de03938..fe12708 100644
--- a/drivers/acpi/osl.c
+++ b/drivers/acpi/osl.c
@@ -1610,7 +1610,7 @@ __setup("acpi_no_auto_serialize", acpi_no_auto_serialize_setup);
 #define ENFORCE_RESOURCES_LAX    1
 #define ENFORCE_RESOURCES_NO     0
 
-static unsigned int acpi_enforce_resources = ENFORCE_RESOURCES_STRICT;
+static unsigned int acpi_enforce_resources = ENFORCE_RESOURCES_NO;
 
 static int __init acpi_enforce_resources_setup(char *str)
 {
