Fix issue that flashcp progress cannot reach 100%

diff --git a/miscutils/flashcp.c b/miscutils/flashcp.c
index 790f9c0..c93fb62 100644
--- a/miscutils/flashcp.c
+++ b/miscutils/flashcp.c
@@ -123,12 +123,14 @@ int flashcp_main(int argc UNUSED_PARAM, char **argv)
 		done = 0;
 		count = BUFSIZE;
 		while (1) {
-			uoff_t rem = statb.st_size - done;
+			uoff_t rem;
+
+			progress(i, done / 1024, (uoff_t)statb.st_size / 1024);
+			rem = statb.st_size - done;
 			if (rem == 0)
 				break;
 			if (rem < BUFSIZE)
 				count = rem;
-			progress(i, done / 1024, (uoff_t)statb.st_size / 1024);
 			xread(fd_f, buf, count);
 			if (i == 1) {
 				int ret;
