Enlarge the delay time of writing 2 bytes

diff --git a/miscutils/24cXX.c b/miscutils/24cXX.c
index b448761..7f8c075 100644
--- a/miscutils/24cXX.c
+++ b/miscutils/24cXX.c
@@ -44,7 +44,7 @@ static int i2c_write_2b(struct eeprom *e, __u8 buf[2])
 	r = i2c_smbus_write_byte_data(e->fd, buf[0], buf[1]);
 	if(r < 0)
 		fprintf(stderr, "Error i2c_write_2b: %s\n", strerror(errno));
-	usleep(1000);
+	usleep(5000);
 	return r;
 }
 
