diff --git a/miscutils/sys_eeprom_mtd.c b/miscutils/sys_eeprom_mtd.c
index 119d77c..aac2e96 100644
--- a/miscutils/sys_eeprom_mtd.c
+++ b/miscutils/sys_eeprom_mtd.c
@@ -23,6 +23,7 @@ int read_sys_eeprom(void *eeprom_data, int offset, int len)
 	return -1;
     }
 
+    /*
     rc = ioctl(fd, MEMGETINFO, &mtdinfo);
 
     if (rc < 0) {
@@ -34,6 +35,7 @@ int read_sys_eeprom(void *eeprom_data, int offset, int len)
         fprintf (stderr, "Unsupported flash type %u\n", mtdinfo.type);
         return -1;
     }
+    */
 
     if (lseek (fd, offset + SYS_EEPROM_OFFSET, SEEK_SET) == -1) {
 	fprintf (stderr, "Seek error on %s: %s\n",
@@ -78,6 +80,7 @@ int write_sys_eeprom(void *eeprom_data, int len)
 	return -1;
     }
 
+    /*
     rc = ioctl(fd, MEMGETINFO, &mtdinfo);
 
     if (rc < 0) {
@@ -89,19 +92,24 @@ int write_sys_eeprom(void *eeprom_data, int len)
         fprintf (stderr, "Unsupported flash type %u\n", mtdinfo.type);
         return -1;
     }
+    */
 
     /*
      * erase entire sys_eeprom
      */
+
+    /*
     erase.length = mtdinfo.size;
     erase.start = 0;
     ioctl (fd, MEMUNLOCK, &erase);
 
+
     if (ioctl (fd, MEMERASE, &erase) != 0) {
 	fprintf (stderr, "MTD erase error on %s: %s\n",
 		SYS_EEPROM_MTD_DEVICE, strerror (errno));
 	return -1;
     }
+    */
 
     if (lseek (fd, SYS_EEPROM_OFFSET, SEEK_SET) == -1) {
 	fprintf (stderr, "Seek error on %s: %s\n",
