From 71d6720577bbc1a6a8be7143f6b50056c3b81add Mon Sep 17 00:00:00 2001
From: "jason.ht.huang" <jason.ht.huang@delta.com.tw>
Date: Tue, 9 Jan 2018 09:56:25 +0800
Subject: [PATCH] i2ctools sys-eeprom delay time

---
 sys_eeprom/24cXX.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/sys_eeprom/24cXX.c b/sys_eeprom/24cXX.c
index be194ad..16927cf 100644
--- a/sys_eeprom/24cXX.c
+++ b/sys_eeprom/24cXX.c
@@ -33,7 +33,7 @@ static int i2c_write_1b(struct eeprom *e, __u8 buf)
 	r = i2c_smbus_write_byte(e->fd, buf);
 	if(r < 0)
 		fprintf(stderr, "Error i2c_write_1b: %s\n", strerror(errno));
-	usleep(10);
+	usleep(5000);
 	return r;
 }
 
@@ -44,7 +44,7 @@ static int i2c_write_2b(struct eeprom *e, __u8 buf[2])
 	r = i2c_smbus_write_byte_data(e->fd, buf[0], buf[1]);
 	if(r < 0)
 		fprintf(stderr, "Error i2c_write_2b: %s\n", strerror(errno));
-	usleep(1000);
+	usleep(5000);
 	return r;
 }
 
@@ -56,7 +56,7 @@ static int i2c_write_3b(struct eeprom *e, __u8 buf[3])
 	r = i2c_smbus_write_word_data(e->fd, buf[0], buf[2] << 8 | buf[1]);
 	if(r < 0)
 		fprintf(stderr, "Error i2c_write_3b: %s\n", strerror(errno));
-	usleep(10);
+	usleep(5000);
 	return r;
 }
 
-- 
2.4.3

