From c36aef27a1546861fefa09aa57910760ebb9cfa2 Mon Sep 17 00:00:00 2001
From: jason <jason@delta.com.tw>
Date: Thu, 21 Dec 2017 03:42:19 +0800
Subject: [PATCH] Modified the string of the type code 0x29

---
 sys_eeprom/onie_tlvinfo.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/sys_eeprom/onie_tlvinfo.h b/sys_eeprom/onie_tlvinfo.h
index 5714f37..8c37b38 100644
--- a/sys_eeprom/onie_tlvinfo.h
+++ b/sys_eeprom/onie_tlvinfo.h
@@ -87,7 +87,7 @@ static const struct tlv_code_desc tlv_code_list[] = {
     { TLV_CODE_DEVICE_VERSION    , "Device Version"},
     { TLV_CODE_LABEL_REVISION    , "Label Revision"},
     { TLV_CODE_PLATFORM_NAME     , "Platform Name"},
-    { TLV_CODE_ONIE_VERSION	 , "Loader Version"},
+    { TLV_CODE_ONIE_VERSION	 , "ONIE Version"},
     { TLV_CODE_MAC_SIZE	         , "MAC Addresses"},
     { TLV_CODE_MANUF_NAME	 , "Manufacturer"},
     { TLV_CODE_MANUF_COUNTRY     , "Country Code"},
-- 
1.9.1

