From 58b1470e7d9ca10f5e239dc3ed2252dba3994833 Mon Sep 17 00:00:00 2001
From: "zoe.wu" <zoe.wu@deltaww.com>
Date: Wed, 22 Aug 2018 11:10:00 +0800
Subject: [PATCH] Patch to change decode onie device version tlv value.

---
 sys_eeprom/onie_tlvinfo.c | 13 ++-----------
 1 file changed, 2 insertions(+), 11 deletions(-)

diff --git a/sys_eeprom/onie_tlvinfo.c b/sys_eeprom/onie_tlvinfo.c
index 84cfc53..4c05f4a 100755
--- a/sys_eeprom/onie_tlvinfo.c
+++ b/sys_eeprom/onie_tlvinfo.c
@@ -265,7 +265,7 @@ static void decode_tlv_value(tlvinfo_tlv_t * tlv, char* value)
 		tlv->value[3], tlv->value[4], tlv->value[5]);
 	break;
     case TLV_CODE_DEVICE_VERSION:
-	sprintf(value, "%u", tlv->value[0]);
+    sprintf(value, "0x%02X (ASCII=%c)", tlv->value[0], tlv->value[0]);
 	break;
     case TLV_CODE_MAC_SIZE:
 	sprintf(value, "%u", (tlv->value[0] << 8) | tlv->value[1]);
@@ -644,19 +644,10 @@ bool tlvinfo_add_tlv(u_int8_t * eeprom, int tcode, char * strval)
     case TLV_CODE_VENDOR_NAME:
     case TLV_CODE_DIAG_VERSION:
     case TLV_CODE_SERVICE_TAG:
+    case TLV_CODE_DEVICE_VERSION:
 	strncpy(data, strval, MAX_TLV_VALUE_LEN);
 	new_tlv_len = min(MAX_TLV_VALUE_LEN, strlen(strval));
 	break;
-    case TLV_CODE_DEVICE_VERSION:
-	value = strtoul(strval, NULL, 0);
-	if (value >= 256) {
-	    printf("ERROR: Device version must be 255 or less. Value " \
-		   "supplied: %u", value);
-	    return(FALSE);
-	}
-	data[0] = value & 0xFF;
-	new_tlv_len = 1;
-	break;
     case TLV_CODE_MAC_SIZE:
 	value = strtoul(strval, NULL, 0);
 	if (value >= 65536) {
-- 
2.4.3

