From c767eeeafcf7dc37c376c81a3b72263075577d77 Mon Sep 17 00:00:00 2001
From: "zoe.wu" <zoe.wu@deltaww.com>
Date: Mon, 8 Oct 2018 10:02:46 +0800
Subject: [PATCH] i2ctools sys-eeprom device version

---
 sys_eeprom/onie_tlvinfo.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/sys_eeprom/onie_tlvinfo.c b/sys_eeprom/onie_tlvinfo.c
index c2c0b1b..adffb39 100755
--- a/sys_eeprom/onie_tlvinfo.c
+++ b/sys_eeprom/onie_tlvinfo.c
@@ -265,7 +265,7 @@ static void decode_tlv_value(tlvinfo_tlv_t * tlv, char* value)
 		tlv->value[3], tlv->value[4], tlv->value[5]);
 	break;
     case TLV_CODE_DEVICE_VERSION:
-	sprintf(value, "%u", tlv->value[0]);
+	sprintf(value, "%u (%02X)", tlv->value[0], tlv->value[0]);
 	break;
     case TLV_CODE_MAC_SIZE:
 	sprintf(value, "%u", (tlv->value[0] << 8) | tlv->value[1]);
-- 
2.4.3

