Enlarge the delay time of writing 3 bytes

diff --git a/miscutils/24cXX.c b/miscutils/24cXX.c
index b448761..c3312c5 100644
--- a/miscutils/24cXX.c
+++ b/miscutils/24cXX.c
@@ -56,7 +56,7 @@ static int i2c_write_3b(struct eeprom *e, __u8 buf[3])
 	r = i2c_smbus_write_word_data(e->fd, buf[0], buf[2] << 8 | buf[1]);
 	if(r < 0)
 		fprintf(stderr, "Error i2c_write_3b: %s\n", strerror(errno));
-	usleep(10);
+	usleep(4000);
 	return r;
 }
 
