read mtd file after flash programmed

diff --git a/miscutils/sys_eeprom_mtd.c b/miscutils/sys_eeprom_mtd.c
index 119d77c..00efd88 100644
--- a/miscutils/sys_eeprom_mtd.c
+++ b/miscutils/sys_eeprom_mtd.c
@@ -122,5 +122,12 @@ int write_sys_eeprom(void *eeprom_data, int len)
 	return -1;
     }
 
+#ifndef DEBUG
+    FILE *fp;
+    fp = fopen(SYS_EEPROM_MTD_DEVICE, "r");
+    fgetc(fp);
+    fclose(fp);
+#endif
+
     return ret;
 }
