diff --git a/util-linux/volume_id/get_devname.c b/util-linux/volume_id/get_devname.c
index 6b97df1..c787b8a 100644
--- a/util-linux/volume_id/get_devname.c
+++ b/util-linux/volume_id/get_devname.c
@@ -253,6 +253,10 @@ int add_to_uuid_cache(const char *device)
 #endif
 	int fd;
 
+	/* Skip mmcblk0rpmb for eMMC card */
+	if (strstr(device, "mmcblk0rpmb") != NULL)
+		return 0;
+
 	fd = open(device, O_RDONLY);
 	if (fd < 0)
 		return 0;
