diff -Nur a/Make.defaults b/Make.defaults
--- a/Make.defaults	2016-04-12 11:58:48.279839798 +0800
+++ b/Make.defaults	2016-04-12 11:56:32.541923997 +0800
@@ -12,15 +12,15 @@
 
 ARCH = $(shell uname -m)
 clang_cflags =
-gcc_cflags = -Wmaybe-uninitialized
+gcc_cflags =
 cflags	:= $(CFLAGS) \
-	-Wall -Wsign-compare -std=gnu11 -fshort-wchar -fPIC \
+	-Wall -Wsign-compare -std=gnu1x -fshort-wchar -fPIC \
 	-fvisibility=hidden \
 	-D_GNU_SOURCE -I${TOPDIR}/src/include/efivar/ \
 	$(if $(filter $(CC),clang),$(clang_cflags),) \
 	$(if $(filter $(CC),gcc),$(gcc_cflags),)
 host_cflags	:= $(HOST_CFLAGS) \
-	-Wall -Wsign-compare -std=gnu11 -fshort-wchar -fPIC \
+	-Wall -Wsign-compare -std=gnu1x -fshort-wchar -fPIC \
 	-fvisibility=hidden \
 	-D_GNU_SOURCE -I${TOPDIR}/src/include/efivar/ \
 	$(if $(filter $(HOST_CC),clang),$(clang_cflags),) \
diff -Nur a/src/Makefile b/src/Makefile
--- a/src/Makefile	2016-04-12 12:01:37.689512485 +0800
+++ b/src/Makefile	2016-04-12 11:56:32.541923997 +0800
@@ -55,10 +55,10 @@
 include/efivar/efivar.h : include/efivar/efivar-guids.h
 
 fakeguid.o : guid.c
-	$(HOST_CC) $(host_cflags) -DEFIVAR_BUILD_ENVIRONMENT -c -o $@ $^
+	$(HOST_CC) $(host_cflags) -DEFIVAR_BUILD_ENVIRONMENT -c -o $@ $<
 
 makeguids.o : makeguids.c
-	$(HOST_CC) $(host_cflags) -DEFIVAR_BUILD_ENVIRONMENT -c -o $@ $^
+	$(HOST_CC) $(host_cflags) -DEFIVAR_BUILD_ENVIRONMENT -c -o $@ $<
 
 makeguids : makeguids.o fakeguid.o
 	$(HOST_CC) $(host_cflags) -o $@ $^ -ldl
