

diff --git a/arch/arm/boot/dts/dni_3048up.dts b/arch/arm/boot/dts/dni_3048up.dts
new file mode 100755
index 0000000..3215058
--- /dev/null
+++ b/arch/arm/boot/dts/dni_3048up.dts
@@ -0,0 +1,237 @@
+/*
+ *	AMZ3048UP Device Tree Source
+ */
+/dts-v1/;
+
+/ {
+        #address-cells = <0x1>;
+        #size-cells = <0x1>;
+        interrupt-parent = <0x1>;
+        model = "dni,3048up";
+        compatible = "dni,3048up";
+        chosen {
+        };
+        aliases {
+                serial0 = "/chipcommonA/serial@0300";
+        };
+        memory {
+                device_type = "memory";
+                reg = <0x61000000 0x3f000000>;
+        };
+        chipcommonA {
+                compatible = "simple-bus";
+                ranges = <0x0 0x18000000 0x1000>;
+                #address-cells = <0x1>;
+                #size-cells = <0x1>;
+                serial@0300 {
+                        compatible = "ns16550";
+                        reg = <0x300 0x100>;
+                        interrupts = <0x7b>;
+                        clock-frequency = <0x5f5e100>;
+                        status = "enabled";
+                };
+        };
+        mpcore {
+                compatible = "simple-bus";
+                ranges = <0x0 0x19020000 0x3000>;
+                #address-cells = <0x1>;
+                #size-cells = <0x1>;
+                scu@0000 {
+                        compatible = "arm,cortex-a9-scu";
+                        reg = <0x0 0x100>;
+                };
+                interrupt-controller@1000 {
+                        compatible = "arm,cortex-a9-gic";
+                        #interrupt-cells = <0x3>;
+                        #address-cells = <0x0>;
+                        interrupt-controller;
+                        reg = <0x1000 0x1000 0x100 0x100>;
+                        linux,phandle = <0x1>;
+                        phandle = <0x1>;
+                };
+                cache-controller@2000 {
+                        compatible = "arm,pl310-cache";
+                        reg = <0x2000 0x1000>;
+                        cache-unified;
+                        cache-level = <0x2>;
+                        linux,phandle = <0x2>;
+                        phandle = <0x2>;
+                };
+        };
+        clocks {
+                #address-cells = <0x1>;
+                #size-cells = <0x0>;
+                periph {
+                        compatible = "fixed-clock";
+                        #clock-cells = <0x0>;
+                        clock-frequency = <0x17d78400>;
+                };
+        };
+        bcm_dma {
+                compatible = "early-dma-alloc";
+                region_size = <0x4000000>;
+                alignment = <0x100000>;
+        };
+        cpus {
+                #address-cells = <0x1>;
+                #size-cells = <0x0>;
+                cpu@0 {
+                        device_type = "cpu";
+                        compatible = "arm,cortex-a9";
+                        next-level-cache = <0x2>;
+                        reg = <0x0>;
+                };
+                cpu@1 {
+                        device_type = "cpu";
+                        compatible = "arm,cortex-a9";
+                        next-level-cache = <0x2>;
+                        reg = <0x1>;
+                };
+        };
+                
+        localbus@1e000000 {
+	        #address-cells = <0x2>;
+	        #size-cells = <0x1>;
+	        compatible = "simple-bus";
+	        ranges = <0x0 0x0 0x1e000000 0x02000000>;
+	
+	        spi@0,0 {
+	            #address-cells = <0x1>;
+	            #size-cells = <0x1>;
+	            compatible = "m25p80";
+	            reg = <0x0 0x0 0x02000000>;
+	            bank-width = <0x1>;
+	            device-width = <0x1>;
+	
+	            partition@0 {
+	                /* uboot - 1MB*/
+	                reg = <0x00000000 0x00100000>;
+	                label = "uboot";
+	            };
+	            partition@1 {
+	                /* uboot-env - 128KB*/
+	                reg = <0x00100000 0x00020000>;
+	                label = "uboot-env";
+	                env_size = <0x10000>;
+	            };
+	            partition@2 {
+	                /* shmoo - 256KB */
+	                reg = <0x00120000 0x00040000>;
+	                label = "shmoo";
+	            };
+	            partition@3 {
+	                /* sys-eeprom - 64KB */
+	                reg = <0x00160000 0x00010000>;
+	                label = "sys-eeprom";
+	            };
+	            partition@4 {
+	                /* onie - 14912KB */
+	                reg = <0x00170000 0x00e90000>;
+	                label = "onie";
+	            };
+	            partition@5 {
+	                /* diag - 16MB */
+	                reg = <0x01000000 0x01000000>;
+	                label = "diag";
+	            };
+	        };
+	    };
+
+        soc@e00000000 {
+                #address-cells = <0x1>;
+                #size-cells = <0x1>;
+                device_type = "soc";
+                ranges = <0x0 0x0 0xe0000000 0x100000>;
+                i2c@3100 {
+                        #address-cells = <0x1>;
+                        #size-cells = <0x0>;
+                        cell-index = <0x1>;
+                        reg = <0x3100 0x100>;
+                        interrupts = <0x2b 0x2>;
+                        dfsrrvel-cache;
+                        clock-frequency = <0x61a80>;
+                        mux@2 {
+                                compatible = "dni,6448-i2c-mux";
+                                reg = <0x2>;
+                                #address-cells = <0x1>;
+                                #size-cells = <0x0>;
+                                i2c@0 {
+                                        #address-cells = <0x1>;
+                                        #size-cells = <0x0>;
+                                        reg = <0x0>;
+                                        tmon@48 {
+                                                compatible = "ti,tmp75";
+                                                reg = <0x48>;
+                                        };
+                                        tmon@49 {
+                                                compatible = "ti,tmp75";
+                                                reg = <0x49>;
+                                        };
+                                };
+                                i2c@1 {
+                                        #address-cells = <0x1>;
+                                        #size-cells = <0x0>;
+                                        reg = <0x1>;
+                                };
+                                i2c@2 {
+                                        #address-cells = <0x1>;
+                                        #size-cells = <0x0>;
+                                        reg = <0x2>;
+                                        fan@48 {
+                                                compatible = "maxim,max6651";
+                                                reg = <0x48>;
+                                        };
+                                        rtc@68 {
+                                                compatible = "stm,m41st85";
+                                                reg = <0x68>;
+                                        };
+                                };
+                                i2c@3 {
+                                        #address-cells = <0x1>;
+                                        #size-cells = <0x0>;
+                                        reg = <0x3>;
+                                };
+                                i2c@4 {
+                                        #address-cells = <0x1>;
+                                        #size-cells = <0x0>;
+                                        reg = <0x4>;
+                                        sfp_eeprom@50 {
+                                                compatible = "at,24c04";
+                                                reg = <0x50>;
+                                                label = "port49";
+                                        };
+                                };
+                                i2c@5 {
+                                        #address-cells = <0x1>;
+                                        #size-cells = <0x0>;
+                                        reg = <0x5>;
+                                        sfp_eeprom@50 {
+                                                compatible = "at,24c04";
+                                                reg = <0x50>;
+                                                label = "port50";
+                                        };
+                                };
+                                i2c@6 {
+                                        #address-cells = <0x1>;
+                                        #size-cells = <0x0>;
+                                        reg = <0x6>;
+                                        sfp_eeprom@50 {
+                                                compatible = "at,24c04";
+                                                reg = <0x50>;
+                                                label = "port51";
+                                        };
+                                };
+                                i2c@7 {
+                                        #address-cells = <0x1>;
+                                        #size-cells = <0x0>;
+                                        reg = <0x7>;
+                                        sfp_eeprom@50 {
+                                                compatible = "at,24c04";
+                                                reg = <0x50>;
+                                                label = "port52";
+                                        };
+                                };
+                        };
+                };
+        };
+};
\ No newline at end of file
diff --git a/arch/arm/mach-iproc/Kconfig b/arch/arm/mach-iproc/Kconfig
old mode 100644
new mode 100755
index ed24988..65fda7a
--- a/arch/arm/mach-iproc/Kconfig
+++ b/arch/arm/mach-iproc/Kconfig
@@ -41,6 +41,11 @@ config MACH_GH
 	help
 	  Support for the Broadcom Greyhound bring-up board.
 
+config MACH_DNI_3048UP
+	bool "Support DNI 3048UP board"
+	help
+	  Support for the DNI 3048UP board.
+
 endchoice
 
 config MACH_IPROC_P7
diff --git a/arch/arm/mach-iproc/board_bu.c b/arch/arm/mach-iproc/board_bu.c
old mode 100644
new mode 100755
index b45588d..1183232
--- a/arch/arm/mach-iproc/board_bu.c
+++ b/arch/arm/mach-iproc/board_bu.c
@@ -379,7 +379,7 @@ static struct resource fa2_resources[] = {
 #ifdef CONFIG_IPROC_I2C
 /* SMBus (I2C/BSC) block */
 
-#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2)
+#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT22) || defined(CONFIG_MACH_DNI_3048UP)
 /* Helix4 */
 static struct resource smbus_resources[] = {
     [0] = {
@@ -447,7 +447,8 @@ static struct platform_device board_smbus_device = {
     .num_resources = ARRAY_SIZE(smbus_resources),
     .resource = smbus_resources,
 };
-#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2)
+#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || \
+	defined(CONFIG_MACH_DNI_3048UP)
 static struct platform_device board_smbus_device1 = {
     .name= "iproc-smb",
     .id = 1,
@@ -476,6 +477,30 @@ static struct platform_device board_fa2_device = {
 #endif
 #endif /* CONFIG_IPROC_FA2 */
 
+#if defined(CONFIG_MACH_DNI_3048UP) /* DNI MFG use following */
+//////////////////////////////////////////////////////////////
+/* DNI_MFG START */
+static struct i2c_board_info __initdata mfg_smbus_devices[] = {
+    { I2C_BOARD_INFO("SFP"               , 0xA0>>1), },
+    { I2C_BOARD_INFO("RTC"               , 0xD0>>1), },
+
+    { I2C_BOARD_INFO("AVS_PWR"           , 0x20>>1), },
+    { I2C_BOARD_INFO("TMP75_MAC"         , 0x92>>1), },
+    { I2C_BOARD_INFO("TMP75_PHY"         , 0x94>>1), },
+    { I2C_BOARD_INFO("CPLD"              , 0x50>>1), },
+};
+
+static struct i2c_board_info __initdata mfg_smbus1_devices[] = {
+    { I2C_BOARD_INFO("FAN"               , 0x58>>1), },
+    { I2C_BOARD_INFO("CURRENT_MONITOR_IC", 0x80>>1), },
+    { I2C_BOARD_INFO("PS_1_EEPROM"       , 0xA0>>1), },
+    { I2C_BOARD_INFO("PS_2_EEPROM"       , 0xA2>>1), },
+    { I2C_BOARD_INFO("POE_PS_EEPROM"     , 0xA4>>1), },
+    { I2C_BOARD_INFO("PSU_FAN_CONTROL"   , 0xB0>>1), }, /* rev1.1 to 0xB0 */
+    { I2C_BOARD_INFO("POE_MAIN_PWR"      , 0xA8>>1), }, /* rev1.1 to 0xA8 */
+};
+#endif /* DNI */
+
 #ifdef CONFIG_IPROC_USB3H
 static struct resource bcm_xhci_resources[] = {
         [0] = {
@@ -516,8 +541,10 @@ static struct resource usbh_ehci_resource[] = {
 		.flags = IORESOURCE_MEM,
 	},
 	[1] = {
-		.start = BCM_INT_ID_USB2H2CORE_USB2_INT,
-		.end = BCM_INT_ID_USB2H2CORE_USB2_INT,
+		.start = 117,
+		.end = 117,
+		//.start = BCM_INT_ID_USB2H2CORE_USB2_INT,
+		//.end = BCM_INT_ID_USB2H2CORE_USB2_INT,
 		.flags = IORESOURCE_IRQ,
 	},
 };
@@ -546,8 +573,10 @@ static struct resource usbh_ohci_resource[] = {
 		.flags = IORESOURCE_MEM,
 	},
 	[1] = {
-		.start = BCM_INT_ID_USB2H2CORE_USB2_INT,
-		.end = BCM_INT_ID_USB2H2CORE_USB2_INT,
+		.start = 117,
+		.end = 117,
+		//.start = BCM_INT_ID_USB2H2CORE_USB2_INT,
+		//.end = BCM_INT_ID_USB2H2CORE_USB2_INT,
 		.flags = IORESOURCE_IRQ,
 	},
 };
@@ -886,7 +915,7 @@ static struct platform_device *board_sata_device[] __initdata = {
 static struct platform_device *board_devices[] __initdata = {
 #ifdef CONFIG_IPROC_I2C
     &board_smbus_device,
-#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2)
+#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP)
     &board_smbus_device1,
 #endif
 #endif /* CONFIG_IPROC_I2C */
@@ -974,7 +1003,6 @@ struct sys_timer board_timer = {
 	.init   = board_timer_init,
 };
 
-
 void __init board_init(void)
 {
 	uint32_t	sata_enable=0;
@@ -1039,6 +1067,11 @@ void __init board_init(void)
 	spi_register_board_info(bcm5301x_spi_device,
 	ARRAY_SIZE(bcm5301x_spi_device));
 
+#if defined(CONFIG_MACH_DNI_3048UP)
+	i2c_register_board_info(0, mfg_smbus_devices, ARRAY_SIZE(mfg_smbus_devices));
+	i2c_register_board_info(1, mfg_smbus1_devices, ARRAY_SIZE(mfg_smbus1_devices));
+#endif
+
 	printk(KERN_DEBUG "board_init: Leave\n");
 }
 
@@ -1056,6 +1089,11 @@ MACHINE_START(IPROC, "Broadcom iProc")
 	.init_machine = board_init,
 MACHINE_END
 
+static const char * helix4_dt_board_compat[] = {
+    "dni,3048up",
+    NULL
+};
+
 DT_MACHINE_START(HELIX4_DT, "Broadcom Helix4 (Flattened Device Tree)")
 	.map_io     = board_map_io,
 	.init_early = iproc_init_early,
diff --git a/arch/arm/mach-iproc/common.c b/arch/arm/mach-iproc/common.c
old mode 100644
new mode 100755
index 132d8fa..a8805e8
--- a/arch/arm/mach-iproc/common.c
+++ b/arch/arm/mach-iproc/common.c
@@ -45,7 +45,8 @@
 
 #if (defined(CONFIG_MACH_NS) || defined(CONFIG_MACH_NSP))
 #define IRQ_IPROC_UART0  117
-#elif (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#elif (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || \
+        defined(CONFIG_MACH_DNI_3048UP))
 #define IRQ_IPROC_UART0  123
 #elif defined(CONFIG_MACH_HR2)
 #define IRQ_IPROC_UART0  123
diff --git a/arch/arm/mach-iproc/flash.c b/arch/arm/mach-iproc/flash.c
old mode 100644
new mode 100755
index 5eaf3ec..1866990
--- a/arch/arm/mach-iproc/flash.c
+++ b/arch/arm/mach-iproc/flash.c
@@ -44,24 +44,34 @@
 #if defined(CONFIG_IPROC_QSPI) || defined(CONFIG_IPROC_QSPI_MODULE)
 static struct mtd_partition sflash_partition_map[] = {
     {
-        .name = "boot",
+        .name = "uboot",
         .offset = 0x00000000,
-        .size = 640 * 1024,
+        .size = 0x00100000,
     },
     {
-        .name = "env",
+        .name = "uboot-env",
         .offset = MTDPART_OFS_APPEND,
-        .size = 384 * 1024,
+        .size = 0x00020000,
     },
     {
-        .name = "system",
+        .name = "shmoo",
         .offset = MTDPART_OFS_APPEND,
-        .size = 15 * 1024 * 1024,
+        .size = 0x00040000,
     },
     {
-        .name = "rootfs",
+        .name = "sys-eeprom",
         .offset = MTDPART_OFS_APPEND,
-        .size = MTDPART_SIZ_FULL,
+        .size = 0x00010000,
+    },
+    {
+        .name = "onie",
+        .offset = MTDPART_OFS_APPEND,
+        .size = 0x00e90000,
+    },
+    {
+        .name = "diag",
+        .offset = MTDPART_OFS_APPEND,
+        .size = 0x01000000,
     },
 };
 #endif /* CONFIG_IPROC_QSPI || CONFIG_IPROC_QSPI_MODULE */
diff --git a/arch/arm/mach-iproc/include/mach/iproc_regs.h b/arch/arm/mach-iproc/include/mach/iproc_regs.h
old mode 100644
new mode 100755
index 935f579..fb84634
--- a/arch/arm/mach-iproc/include/mach/iproc_regs.h
+++ b/arch/arm/mach-iproc/include/mach/iproc_regs.h
@@ -23,7 +23,7 @@
 #elif defined(CONFIG_MACH_NS)
 #include "socregs_ns_open.h"
 #elif (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || \
-	defined(CONFIG_MACH_HR2))
+	defined(CONFIG_MACH_HR2) || defined(CONFIG_MACH_DNI_3048UP))
 #include "socregs_ing_open.h"
 #elif defined(CONFIG_MACH_NSP)
 #include "socregs_nsp_open.h"
@@ -98,7 +98,8 @@
 
 #endif
 
-#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_NSP) || defined(CONFIG_MACH_KT2)
+#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_NSP) || \
+		defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP)
 #define CCB_RNG_CTRL    ChipcommonB_rng_CTRL
 #endif
 
@@ -332,7 +333,8 @@
 
 #if defined(CONFIG_MACH_NS)
 #define IPROC_USB20_REG_BASE		  (0x18021000)
-#elif defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_NSP) || defined(CONFIG_MACH_KT2)
+#elif defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_NSP) || \
+			defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP)
 #define IPROC_USB20_REG_BASE		  (0x1802A000)
 #define IPROC_UDC_IRQ				  (238)
 #endif
@@ -340,7 +342,7 @@
 //#define IPROC_USB30_REG_BASE		  (0x18022000)
 #define IPROC_USB20_PHY_REG_BASE	  (0x18023000) /* ??*/
 
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 #define IPROC_USB2D_REG_BASE	USB2D_ENDPNT_IN_CTRL_0
 #define IPROC_USB2D_REG_SIZE	(0x2000) /* 8KB */
 #endif
@@ -355,7 +357,7 @@
 #define IPROC_GMAC1_INT				180
 #define IPROC_GMAC2_INT				181
 #define IPROC_GMAC3_INT				182
-#elif (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#elif (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 #define IPROC_NUM_GMACS				2
 #define IPROC_GMAC0_REG_BASE		  (GMAC0_DEVCONTROL) //(0x18022000)
 #define IPROC_GMAC1_REG_BASE		  (GMAC1_DEVCONTROL) //(0x18023000)
@@ -398,7 +400,7 @@
 #define IPROC_DMU_REG_BASE				(0x1800c000)
 #define IPROC_IDM_REG_BASE				(0x18100000)
 #elif defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_HR2) || defined(CONFIG_MACH_KT2) \
-	|| defined(CONFIG_MACH_GH)
+	|| defined(CONFIG_MACH_GH) || defined(CONFIG_MACH_DNI_3048UP)
 #define IPROC_CRU_REG_BASE				CRU_control
 #define IPROC_DMU_REG_BASE				DMU_PCU_IPROC_CONTROL
 #define IPROC_IDM_REG_BASE				(IHOST_M0_IO_CONTROL_DIRECT - 0x408)
@@ -422,7 +424,7 @@
 #define IPROC_STRAP_NAND_TYPE_SHIFT		(12)
 #define IPROC_STRAP_NAND_PAGE_SHIFT		(10)
 #elif defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_HR2) || defined(CONFIG_MACH_KT2) \
-	|| defined(CONFIG_MACH_GH)
+	|| defined(CONFIG_MACH_GH) || defined(CONFIG_MACH_DNI_3048UP)
 #define IPROC_DMU_STRAPS_OFFSET			DMU_PCU_IPROC_STRAPS_CAPTURED_BASE
 #define IPROC_STRAP_BOOT_DEV_SHIFT		DMU_PCU_IPROC_STRAPS_CAPTURED__strap_boot_dev_R
 #define IPROC_STRAP_NAND_TYPE_SHIFT		DMU_PCU_IPROC_STRAPS_CAPTURED__strap_nand_type_R
@@ -445,7 +447,8 @@
 #define IPROC_IDM_QSPI_REG_BASE		  QSPI_IDM_IDM_IO_CONTROL_DIRECT
 #define IPROC_QSPI_IRQ_START		  (104)
 #define IPROC_QSPI_IRQ_END		  (109)
-#elif defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_HR2) || defined(CONFIG_MACH_KT2)
+#elif defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_HR2) || \
+			defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP)
 #define IPROC_IDM_NAND_REG_BASE		  NAND_IDM_IDM_IO_CONTROL_DIRECT
 #define IPROC_NAND_IRQ_START		  (106)
 #define IPROC_IDM_QSPI_REG_BASE		  QSPI_IDM_IDM_IO_CONTROL_DIRECT
@@ -620,7 +623,7 @@
 #define IPROC_GPIO_CCA_CTRL0            (0x01c0)
 
 #else
-/* CONFIG_MACH_HX4, CONFIG_MACH_HR2, CONFIG_MACH_NSP, CONFIG_MACH_KT2 */
+/* CONFIG_MACH_HX4, CONFIG_MACH_HR2, CONFIG_MACH_NSP, CONFIG_MACH_KT2, CONFIG_MACH_DNI_3048UP */
 #define IPROC_CCA_INT_STS               (ChipcommonA_IntStatus_BASE)
 #define IPROC_CCA_INT_MASK              (ChipcommonA_IntMask_BASE)
 #define IPROC_GPIO_CCA_BASE             (ChipcommonA_GPIOInput)
@@ -691,7 +694,7 @@
 #define IPROC_GPIO_CCB_PRB_EN       (ChipcommonG_GP_PRB_ENABLE_BASE)
 #define IPROC_GPIO_CCB_PRB_OE       (ChipcommonG_GP_PRB_OE_BASE)
 #else
-/* CONFIG_MACH_HX4, CONFIG_MACH_HR2, CONFIG_MACH_NSP, CONFIG_MACH_KT2 */
+/* CONFIG_MACH_HX4, CONFIG_MACH_HR2, CONFIG_MACH_NSP, CONFIG_MACH_KT2, CONFIG_MACH_DNI_3048UP */
 #define IPROC_GPIO_CCB_BASE         (ChipcommonB_GP_DATA_IN)
 #define IPROC_GPIO_CCB_DIN          (ChipcommonB_GP_DATA_IN_BASE)
 #define IPROC_GPIO_CCB_DOUT         (ChipcommonB_GP_DATA_OUT_BASE)
@@ -719,7 +722,7 @@
 #elif defined(CONFIG_MACH_IPROC_P7)
 #define IPROC_GPIO_CCG_INT          (116)
 #else
-/* CONFIG_MACH_HX4, CONFIG_MACH_HR2, CONFIG_MACH_KT2 */
+/* CONFIG_MACH_HX4, CONFIG_MACH_HR2, CONFIG_MACH_KT2, CONFIG_MACH_DNI_3048UP */
 #define IPROC_GPIO_CCA_INT          (123)
 #define IPROC_GPIO_CCB_INT          (125)
 #endif
@@ -772,7 +775,7 @@
 #define IPROC_CCB_TIMER3_REGS_VA    (IPROC_CCB_TIM1_REG_VA + 0x20)
 #define IPROC_CCB_TIMER_INT_START   (122)
 #define IPROC_CCB_TIMER_INT_COUNT   (4)
-#else /* CONFIG_MACH_HX4, CONFIG_MACH_HR2, CONFIG_MACH_KT2 */
+#else /* CONFIG_MACH_HX4, CONFIG_MACH_HR2, CONFIG_MACH_KT2, CONFIG_MACH_DNI_3048UP */
 #define IPROC_CCB_TIMER0_REGS_VA    (IPROC_CCB_TIM0_REG_VA + 0x00)
 #define IPROC_CCB_TIMER1_REGS_VA    (IPROC_CCB_TIM0_REG_VA + 0x20)
 #define IPROC_CCB_TIMER2_REGS_VA    (IPROC_CCB_TIM1_REG_VA + 0x00)
@@ -782,10 +785,14 @@
 #endif
 
 /* ChipCommonB Watchdog */
-/* CCB WDT could be set only when CONFIG_MACH_HR2, CONFIG_MACH_HX4, or CONFIG_MACH_NSP is set */
+/*
+ * CCB WDT could be set only when CONFIG_MACH_HR2, CONFIG_MACH_HX4,
+ * CONFIG_MACH_DNI_3048UP or CONFIG_MACH_NSP is set
+ */
 #define IPROC_CCB_WDT_WDOGLOAD	ChipcommonB_WDT_WDOGLOAD
 #define IPROC_CCB_WDT_REG_BASE	IPROC_CCB_WDT_WDOGLOAD
-#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_HR2) || defined(CONFIG_MACH_KT2)
+#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_HR2) || \
+			defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP)
 #define IPROC_CCB_WDT_BOOTSTATUS		DMU_PCU_CRU_RESET_REASON
 #define IPROC_CCB_WDT_BOOTSTATUS_BIT	DMU_PCU_CRU_RESET_REASON__watchdog_reset
 #elif defined(CONFIG_MACH_NSP)
diff --git a/arch/arm/mach-iproc/include/mach/socregs_ing_open.h b/arch/arm/mach-iproc/include/mach/socregs_ing_open.h
old mode 100644
new mode 100755
index 53c3b98..49cb77a
--- a/arch/arm/mach-iproc/include/mach/socregs_ing_open.h
+++ b/arch/arm/mach-iproc/include/mach/socregs_ing_open.h
@@ -84,7 +84,7 @@
 #define ChipcommonB_PWM_DUTYHI_COUNT3_BASE 0x020
 #define ChipcommonB_PWMCTL_BASE 0x000
 #define ChipcommonB_rng_CTRL 0x18033000
-#if (defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_HX4))
+#if (defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_DNI_3048UP))
 #define USB2_IDM_IDM_IO_CONTROL_DIRECT 0x18115408
 #define USB2D_IDM_IDM_IO_CONTROL_DIRECT 0x18116408
 #define USB2D_IDM_IDM_IO_CONTROL_DIRECT__clk_enable 0
@@ -93,7 +93,7 @@
 #endif
 #define DMU_CRU_RESET_BASE 0x200
 #define ChipcommonB_SMBus1_SMBus_Config 0x1803b000
-#if (defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_HX4))
+#if (defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_DNI_3048UP))
 #define USB2D_ENDPNT_IN_CTRL_0 0x18042000
 #endif
 #define IHOST_S1_IDM_ERROR_LOG_CONTROL 0x18107900
@@ -131,7 +131,7 @@
 #define AXI_PCIE_S0_IDM_IDM_ERROR_LOG_ID 0x1810b914
 #define AXI_PCIE_S0_IDM_IDM_ERROR_LOG_FLAGS 0x1810b91c
 #define AXI_PCIE_S0_IDM_IDM_INTERRUPT_STATUS 0x1810ba00
-#if (defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_HX4))
+#if (defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_DNI_3048UP))
 #define AXI_PCIE_S1_IDM_IDM_ERROR_LOG_CONTROL 0x1810c900
 #define AXI_PCIE_S1_IDM_IDM_ERROR_LOG_COMPLETE 0x1810c904
 #define AXI_PCIE_S1_IDM_IDM_ERROR_LOG_STATUS 0x1810c908
@@ -196,7 +196,7 @@
 #define APBZ_S0_IDM_IDM_ERROR_LOG_ID 0x18121914
 #define APBZ_S0_IDM_IDM_ERROR_LOG_FLAGS 0x1812191c
 #define APBZ_S0_IDM_IDM_INTERRUPT_STATUS 0x18121a00
-#if (defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_HX4))
+#if (defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_DNI_3048UP))
 #define APBV_S0_IDM_IDM_ERROR_LOG_CONTROL 0x18122900
 #define APBV_S0_IDM_IDM_ERROR_LOG_COMPLETE 0x18122904
 #define APBV_S0_IDM_IDM_ERROR_LOG_STATUS 0x18122908
@@ -212,7 +212,7 @@
 #define AXIIC_DS_3_IDM_ERROR_LOG_ID 0x18123914
 #define AXIIC_DS_3_IDM_ERROR_LOG_FLAGS 0x1812391c
 #define AXIIC_DS_3_IDM_INTERRUPT_STATUS 0x18123a00
-#if (defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_HX4))
+#if (defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_DNI_3048UP))
 #define AXIIC_DS_4_IDM_ERROR_LOG_CONTROL 0x18124900
 #define AXIIC_DS_4_IDM_ERROR_LOG_COMPLETE 0x18124904
 #define AXIIC_DS_4_IDM_ERROR_LOG_STATUS 0x18124908
@@ -242,7 +242,7 @@
 #define AXIIC_DS_0_IDM_ERROR_LOG_ID 0x18141914
 #define AXIIC_DS_0_IDM_ERROR_LOG_FLAGS 0x1814191c
 #define AXIIC_DS_0_IDM_INTERRUPT_STATUS 0x18141a00
-#if (defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_HX4))
+#if (defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_DNI_3048UP))
 #define AXIIC_DS_1_IDM_ERROR_LOG_CONTROL 0x18142900
 #define AXIIC_DS_1_IDM_ERROR_LOG_COMPLETE 0x18142904
 #define AXIIC_DS_1_IDM_ERROR_LOG_STATUS 0x18142908
@@ -525,7 +525,7 @@
 #define IPROC_DDR_PLL_STATUS_RDWRMASK 0x00000000
 #define IPROC_DDR_PLL_STATUS_RESETVALUE 0x0
 #endif
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_HR2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_HR2) || defined(CONFIG_MACH_DNI_3048UP))
 #define IPROC_WRAP_USBPHY_CTRL 0x1803fc34
 #define IPROC_WRAP_GEN_PLL_STATUS__GEN_PLL_LOCK 0
 #define IPROC_WRAP_GEN_PLL_CTRL1__NDIV_INT_R 0
@@ -756,7 +756,7 @@
 #define IPROC_WRAP_IPROC_XGPLL_STATUS_DATAMASK 0xffffffff
 #define IPROC_WRAP_IPROC_XGPLL_STATUS_RDWRMASK 0x00000000
 #define IPROC_WRAP_IPROC_XGPLL_STATUS_RESETVALUE 0x0
-#if defined(CONFIG_MACH_HX4)
+#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_DNI_3048UP)
 #define IPROC_WRAP_MISC_CONTROL__QUAD_SERDES_MDIO_SEL 3
 #define IPROC_WRAP_MISC_CONTROL__QUAD_SERDES_CTRL_SEL 2
 #define IPROC_WRAP_MISC_CONTROL__IPROC_MDIO_SEL 4
diff --git a/arch/arm/mach-iproc/northstar.c b/arch/arm/mach-iproc/northstar.c
old mode 100644
new mode 100755
index 112144b..3acb309
--- a/arch/arm/mach-iproc/northstar.c
+++ b/arch/arm/mach-iproc/northstar.c
@@ -88,7 +88,7 @@ static void __init northstar_l2x0_init(void)
 	} else {	/* northstar */
 		l2x0_init(l2cache_base, 0x0A130000, ~(0x000F0000));
 	}
-#elif (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#elif (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 		l2x0_init(l2cache_base, 0x0A150000, ~(0x000F0000));
 #elif defined(CONFIG_MACH_HR2)
 		l2x0_init(l2cache_base, 0x0A120000, ~(0x000F0000));
diff --git a/arch/arm/mach-iproc/northstar_dmu.c b/arch/arm/mach-iproc/northstar_dmu.c
old mode 100644
new mode 100755
index fde19a9..a750d5c
--- a/arch/arm/mach-iproc/northstar_dmu.c
+++ b/arch/arm/mach-iproc/northstar_dmu.c
@@ -23,6 +23,7 @@
 #include <linux/io.h>
 #include <linux/ioport.h>
 #include <linux/clkdev.h>
+#include <linux/i2c.h>
 
 #include <asm/io.h>
 #include <asm/pgtable.h>
@@ -244,7 +245,8 @@ static int genpll_status(struct clk * clk)
 	return 0;
 }
 #endif
-#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_HR2) || defined(CONFIG_MACH_GH)
+#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_HR2) || \
+	defined(CONFIG_MACH_GH) || defined(CONFIG_MACH_DNI_3048UP)
 static int genpll_status(struct clk * clk)
 {
 	u32 reg;
@@ -382,7 +384,8 @@ static int genpll_chan_status(struct clk * clk)
 }
 #endif
 
-#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_HR2) || defined(CONFIG_MACH_GH)
+#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_HR2) || \
+	defined(CONFIG_MACH_GH) || defined(CONFIG_MACH_DNI_3048UP)
 static int genpll_chan_status(struct clk * clk)
 {
 	void * __iomem base;
@@ -696,7 +699,8 @@ static void __init northstar_clocks_init(void *__iomem cru_regs_base,
 #if (defined(CONFIG_MACH_NS) || defined(CONFIG_MACH_NSP))
 	clk_lcpll.regs_base =	cru_regs_base + 0x00 ;
 	clk_genpll.regs_base =	cru_regs_base + 0x40 ;
-#elif defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_HR2) || defined(CONFIG_MACH_GH)
+#elif defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_HR2) || \
+	defined(CONFIG_MACH_GH) || defined(CONFIG_MACH_DNI_3048UP)
 	clk_lcpll.regs_base =	cru_regs_base + 0x1c ;
 	clk_genpll.regs_base =	cru_regs_base + 0x00 ;
 #endif
@@ -729,7 +733,8 @@ void __init northstar_dmu_init(struct clk *clk_ref)
 	/* Initialize clocks */
 #if (defined(CONFIG_MACH_NS) || defined(CONFIG_MACH_NSP))
 	northstar_clocks_init(reg_base + 0x100, clk_ref); /* CRU LCPLL control0 */
-#elif defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_HR2) || defined(CONFIG_MACH_GH)
+#elif defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_HR2) || \
+	defined(CONFIG_MACH_GH) || defined(CONFIG_MACH_DNI_3048UP)
 	northstar_clocks_init(reg_base + 0xc00, clk_ref); /* IPROC_WRAP_GEN_PLL_CTRL0 */
 #elif defined(CONFIG_MACH_KT2)
 	northstar_clocks_init(NULL, clk_ref); /* IPROC_WRAP_GEN_PLL_CTRL0 */
@@ -752,13 +757,19 @@ void northstar_restart(char mode, const char *cmd)
 	reg &= ~((u32) 1 << 1);
 	writel_relaxed(reg, reg_addr);
 #elif defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_HR2) || defined(CONFIG_MACH_KT2) \
-	|| defined(CONFIG_MACH_GH)
+	|| defined(CONFIG_MACH_GH) 
 	reg_addr = (void * __iomem) dmu_regs.start + DMU_CRU_RESET_BASE ;
 	/* Reset iproc and cmicd/switch */
 	writel_relaxed(0, reg_addr);
-#endif
+#elif defined(CONFIG_MACH_DNI_3048UP)
+	struct i2c_adapter *i2c_adap;
+
+	u8 data = 0x7f;
 
+	i2c_adap = i2c_get_adapter(0);
 
+	i2c_smbus_xfer(i2c_adap, 0x28, 0, I2C_SMBUS_WRITE, 0x2, I2C_SMBUS_BYTE_DATA, &data);
+#endif
 }
 
 
diff --git a/arch/arm/plat-iproc/include/mach/irqs.h b/arch/arm/plat-iproc/include/mach/irqs.h
old mode 100644
new mode 100755
index 752ebcf..6f51794
--- a/arch/arm/plat-iproc/include/mach/irqs.h
+++ b/arch/arm/plat-iproc/include/mach/irqs.h
@@ -74,7 +74,7 @@
 #define BCM_INT_ID_IHOST_CT             42
 #define BCM_INT_ID_IHOST_DEFFLG_CPU0    44
 #define BCM_INT_ID_IHOST_DEFFLG_CPU1    45
-#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2)
+#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP)
 #define BCM_INT_ID_IHOST_CPU0_PAR		46
 #define BCM_INT_ID_IHOST_CPU1_PAR	    47
 #define BCM_INT_ID_IHOST_SCU0_PAR		48
@@ -116,7 +116,7 @@
 #define BCM_INT_ID_NAND_S0              (24 + BCM_INT_ID_IHOST_MAX)
 #define BCM_INT_ID_QPSI_S0              (25 + BCM_INT_ID_IHOST_MAX)
 #define BCM_INT_ID_A9JTAG_S0            (26 + BCM_INT_ID_IHOST_MAX)
-#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2)
+#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP)
 #define BCM_INT_ID_SATA_S0				(27 + BCM_INT_ID_IHOST_MAX)
 #define BCM_INT_ID_SRAM_S0              (28+ BCM_INT_ID_IHOST_MAX)
 #define BCM_INT_ID_APBW					(29 + BCM_INT_ID_IHOST_MAX)
@@ -149,7 +149,7 @@
 /* DMAC Interrupt IDs                                                  */
 /*=====================================================================*/
 #define BCM_INT_ID_DMAC                 (0 + BCM_INT_ID_DDR_MAX)
-#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2)
+#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP)
 #define BCM_INT_ID_DMAC_ABORT           (16 + BCM_INT_ID_DDR_MAX)
 #define BCM_INT_ID_DMAC_MAX             (17 + BCM_INT_ID_DDR_MAX)
 #elif defined(CONFIG_MACH_NS) || defined(CONFIG_MACH_NSP)
diff --git a/arch/arm/plat-iproc/platsmp.c b/arch/arm/plat-iproc/platsmp.c
old mode 100644
new mode 100755
index 932847d..739aa7b
--- a/arch/arm/plat-iproc/platsmp.c
+++ b/arch/arm/plat-iproc/platsmp.c
@@ -34,7 +34,7 @@
 #define SOC_ROM_BASE_PA    0xFFFF0000
 #ifdef CONFIG_MACH_NS
 #define SOC_ROM_LUT_OFF    0x400
-#elif defined(CONFIG_MACH_HX4)
+#elif defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_DNI_3048UP)
 #define SOC_ROM_LUT_OFF    0x42c
 #elif defined(CONFIG_MACH_HR2)
 #define SOC_ROM_LUT_OFF    0x400
diff --git a/drivers/bcmdrivers/Makefile b/drivers/bcmdrivers/Makefile
index a58c466..d57a842 100755
--- a/drivers/bcmdrivers/Makefile
+++ b/drivers/bcmdrivers/Makefile
@@ -17,11 +17,11 @@ obj-y += nand/
 obj-y += pwm/
 obj-y += gpio/
 obj-y += wdt/
+obj-y += smbus/
 obj-y += usb2h/
 obj-y += pcie/
 obj-$(CONFIG_ET) += gmac/et/
 obj-y += gmac/hnd/
-obj-y += smbus/
 
 obj-y += pmu/
 
diff --git a/drivers/bcmdrivers/gmac/hnd/Makefile b/drivers/bcmdrivers/gmac/hnd/Makefile
index 44500c7..99869bc 100755
--- a/drivers/bcmdrivers/gmac/hnd/Makefile
+++ b/drivers/bcmdrivers/gmac/hnd/Makefile
@@ -57,6 +57,12 @@ hnd-objs += $(SHARED)/bcmiproc_serdes.o
 HND_OBJS += $(src)/$(SHARED)/bcmiproc_phy5461s.o
 hnd-objs += $(SHARED)/bcmiproc_phy5461s.o
 endif
+ifeq ($(CONFIG_MACH_DNI_3048UP),y)
+HND_OBJS += $(src)/$(SHARED)/bcmiproc_serdes.o
+hnd-objs += $(SHARED)/bcmiproc_serdes.o
+HND_OBJS += $(src)/$(SHARED)/bcmiproc_phy5461s.o
+hnd-objs += $(SHARED)/bcmiproc_phy5461s.o
+endif
 ifeq ($(CONFIG_MACH_KT2),y)
 HND_OBJS += $(src)/$(SHARED)/bcmiproc_serdes.o
 hnd-objs += $(SHARED)/bcmiproc_serdes.o
@@ -103,6 +109,10 @@ ifeq ($(CONFIG_MACH_HX4),y)
 HND_OBJS += $(src)/$(SHARED)/hx4_erom.o
 hnd-objs += $(SHARED)/hx4_erom.o
 endif
+ifeq ($(CONFIG_MACH_DNI_3048UP),y)
+HND_OBJS += $(src)/$(SHARED)/hx4_erom.o
+hnd-objs += $(SHARED)/hx4_erom.o
+endif
 ifeq ($(CONFIG_MACH_HR2),y)
 HND_OBJS += $(src)/$(SHARED)/hr2_erom.o
 hnd-objs += $(SHARED)/hr2_erom.o
diff --git a/drivers/bcmdrivers/gmac/src/et/sys/et_linux.c b/drivers/bcmdrivers/gmac/src/et/sys/et_linux.c
index aa9ad62..52a795e 100755
--- a/drivers/bcmdrivers/gmac/src/et/sys/et_linux.c
+++ b/drivers/bcmdrivers/gmac/src/et/sys/et_linux.c
@@ -220,7 +220,7 @@ extern int fa2_modify_header(struct sk_buff *skb);
 #endif /* defined(CONFIG_IPROC_FA2) */
 
 #if defined(CONFIG_IPROC_SDK_MGT_PORT_HANDOFF)
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 extern int gmac_has_mdio_access(void);
 #endif /* (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2)) */
 #endif /* defined(CONFIG_IPROC_SDK_MGT_PORT_HANDOFF) */
@@ -1453,7 +1453,8 @@ et_init(et_info_t *et, uint options)
 
 	etc_init(et->etc, options);
 
-#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_HR2) || defined(CONFIG_MACH_KT2)
+#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_HR2) || \
+    defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP)
 	netif_carrier_off(et->dev);
 #endif /* defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_HR2) || defined(CONFIG_MACH_KT2) */
 }
@@ -1488,7 +1489,7 @@ et_up(et_info_t *et)
 	etc_up(etc);
 
 #if defined(CONFIG_IPROC_SDK_MGT_PORT_HANDOFF)
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 	if (et->set) {
 		/* This will happen if running watchdog to monitor mdio bus */
 		/* and port not up */
@@ -1544,7 +1545,7 @@ et_down(et_info_t *et, int reset)
 	netif_stop_queue(et->dev);
 
 #if defined(CONFIG_IPROC_SDK_MGT_PORT_HANDOFF)
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 	if (gmac_has_mdio_access()) {
 		/* we have mdio bus don't stop timer so we can continue to monitor */
 		stoptmr = FALSE;
@@ -1609,7 +1610,7 @@ _et_watchdog(struct net_device *dev)
 		add_timer(&et->timer);
 	}
 #if defined(CONFIG_IPROC_SDK_MGT_PORT_HANDOFF)
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 	/* this in case port when up then down before we released mdio */
 	else if (gmac_has_mdio_access()) {
 		/* interface not up but we have mdio bus */
@@ -1694,7 +1695,7 @@ et_get_settings(struct net_device *dev, struct ethtool_cmd *ecmd)
     ecmd->supported = (SUPPORTED_10baseT_Half | SUPPORTED_10baseT_Full |
                        SUPPORTED_100baseT_Half | SUPPORTED_100baseT_Full |
                        SUPPORTED_Autoneg | SUPPORTED_TP);
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
     ecmd->supported |= SUPPORTED_1000baseT_Full | SUPPORTED_Pause;
 #endif
 #if defined(CONFIG_MACH_HR2)
@@ -1716,7 +1717,7 @@ et_get_settings(struct net_device *dev, struct ethtool_cmd *ecmd)
 	        ADVERTISED_1000baseT_Half : 0;
     ecmd->advertising |= (et->etc->forcespeed == ET_AUTO) ?
 	        ADVERTISED_Autoneg : 0;
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_HR2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_HR2) || defined(CONFIG_MACH_DNI_3048UP))
     ecmd->advertising |= ADVERTISED_Pause;
 #endif
     if (et->etc->linkstate) {
@@ -3119,7 +3120,7 @@ if ((osl_ctfpool_init(unit, osh, CTFPOOLSZ, RXBUFSZ+BCMEXTRAHDROOM) < 0)) {
 	et->timer.function = et_watchdog;
 
 #if defined(CONFIG_IPROC_SDK_MGT_PORT_HANDOFF)
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 	/* schedule one second watchdog timer */
 	et->timer.expires = jiffies + HZ;
 	add_timer(&et->timer);
diff --git a/drivers/bcmdrivers/gmac/src/et/sys/etc.c b/drivers/bcmdrivers/gmac/src/et/sys/etc.c
index 00fa88b..33ad943 100755
--- a/drivers/bcmdrivers/gmac/src/et/sys/etc.c
+++ b/drivers/bcmdrivers/gmac/src/et/sys/etc.c
@@ -61,7 +61,7 @@ static void etc_loopback(etc_info_t *etc, int on);
 static void etc_dumpetc(etc_info_t *etc, struct bcmstrbuf *b);
 int etc_gmac_speed(int gmac);
 
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 extern void gmac_set_amac_mdio(int en);
 extern int gmac_has_mdio_access(void);
 #endif /* (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2)) */
@@ -600,7 +600,7 @@ etc_watchdog(etc_info_t *etc)
 		ethup |= 1<<etc->unit;
 	}
 
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 #if defined(CONFIG_IPROC_SDK_MGT_PORT_HANDOFF)
 	if ( !gmac_has_mdio_access()) {
         /* we can't monitor link so force link up */
diff --git a/drivers/bcmdrivers/gmac/src/et/sys/etcgmac.c b/drivers/bcmdrivers/gmac/src/et/sys/etcgmac.c
index cf16543..9c8c2e0 100755
--- a/drivers/bcmdrivers/gmac/src/et/sys/etcgmac.c
+++ b/drivers/bcmdrivers/gmac/src/et/sys/etcgmac.c
@@ -50,13 +50,13 @@
 #ifdef GMAC3
 #include <hndfwd.h>	/* GMAC3 */
 #endif /* GMAC3 */
-#if defined(CONFIG_MACH_HX4)
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_DNI_3048UP))
 #include "mach/socregs_ing_open.h"
 #endif
 #if defined(CONFIG_MACH_KT2)
 #include "mach/socregs_ing_open.h"
 #endif
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 #include "bcmiproc_serdes.h"
 #include "bcmiproc_phy5461s.h"
 #endif
@@ -144,7 +144,7 @@ static void chipphyenable(ch_t *ch, uint eth_num, uint phyaddr, int enable);
 static void chipdumpregs(ch_t *ch, gmacregs_t *regs, struct bcmstrbuf *b);
 static void gmac_mf_cleanup(ch_t *ch);
 static int gmac_speed(ch_t *ch, uint32 speed);
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 static void gmac_serdes_init(ch_t *ch);
 #endif
 static void gmac_miiconfig(ch_t *ch);
@@ -192,12 +192,12 @@ static uint devices[] = {
 	0x0000
 };
 
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 static void *wrapaddr = 0;
 void gmac_set_amac_mdio(int en)
 {
 	u32 tmp;
-#if defined(CONFIG_MACH_HX4)
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_DNI_3048UP))
 	u32 mdio_sel= IPROC_WRAP_MISC_CONTROL__QUAD_SERDES_MDIO_SEL;
 	u32 ctrl_sel= IPROC_WRAP_MISC_CONTROL__QUAD_SERDES_CTRL_SEL;
 #else
@@ -244,7 +244,7 @@ int gmac_has_mdio_access(void)
 	u32 tmp;
 	u32 regmsk = IPROC_WRAP_MISC_CONTROL__IPROC_MDIO_SEL;
 
-#if defined(CONFIG_MACH_HX4)
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_DNI_3048UP))
 	regmsk |= 	IPROC_WRAP_MISC_CONTROL__QUAD_SERDES_MDIO_SEL |
 				IPROC_WRAP_MISC_CONTROL__QUAD_SERDES_CTRL_SEL;
 #else
@@ -439,7 +439,7 @@ chipattach(etc_info_t *etc, void *osh, void *regsva)
 	/* reset the gmac core */
 	chipreset(ch);
 
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 	if (wrapaddr == 0) {
 		/* flip switch so AMAC can access serdes */
 		gmac_set_amac_mdio(1);
@@ -710,7 +710,7 @@ chipattach(etc_info_t *etc, void *osh, void *regsva)
 #endif
 
 	/* reset phy: reset it once now */
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
     if (ch->etc->unit == 0) {
         serdes_reset_core(ch->etc->unit, etc->phyaddr);
     }
@@ -760,7 +760,7 @@ chipattach(etc_info_t *etc, void *osh, void *regsva)
 	}
 #endif /* ETROBO */
 
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
     if (gmac_last_interface(etc->unit)) {
         /* must init all serdes lanes, init port 49 (phy 3) */
         serdes_init(etc->unit, 3);
@@ -1036,7 +1036,7 @@ gmac_reset(ch_t *ch)
 	            CC_CFE | CC_RL | CC_RED | CC_PE | CC_TPI | CC_PAD_EN | CC_PF);
 	cmdcfg |= (CC_PROM | CC_NLC | CC_CFE);
 
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 	cmdcfg |= (CC_AE | CC_OT | CC_OR);
 #endif
 
@@ -1126,7 +1126,7 @@ gmac_speed(ch_t *ch, uint32 speed)
 	cmdcfg &= ~(CC_ES_MASK | CC_HD);
 	cmdcfg |= ((speed << CC_ES_SHIFT) | hd_ena);
 
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 	cmdcfg |= (CC_AE | CC_OT | CC_OR);
 #endif
 
@@ -1253,7 +1253,7 @@ gmac_txflowcontrol(ch_t *ch, bool on)
 	gmac_clear_reset(ch);
 }
 
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 static void
 gmac_serdes_init(ch_t *ch)
 {
@@ -1278,7 +1278,7 @@ gmac_serdes_init(ch_t *ch)
      */
 
 	sdctl = 0;
-#if defined(CONFIG_MACH_HX4)
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_DNI_3048UP))
     sdctl |= (SC_TX1G_FIFO_RST_VAL|SC_FORCE_SPD_STRAP_VAL|SC_REFSEL_VAL|SC_REF_TERM_SEL_MASK);
 #else
 	sdctl |= (SC_TX1G_FIFO_RST_VAL|SC_FORCE_SPD_STRAP_VAL|SC_REF_TERM_SEL_MASK);
@@ -2158,7 +2158,7 @@ chipduplexupd(ch_t *ch)
 		cmdcfg |= duplex;
 	}
 
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 	cmdcfg |= (CC_AE | CC_OT | CC_OR);
 #endif
 
@@ -2251,7 +2251,7 @@ chipphyrd(ch_t *ch, uint phyaddr, uint reg)
 		tmp = 0xffff;
 	}
 #endif /* (defined(CONFIG_MACH_NS) || defined(CONFIG_MACH_NSP)) */
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 	uint32 addr, int_bus, bank, flags;
 	uint16 tmp16;
 
@@ -2317,7 +2317,7 @@ chipphywr(ch_t *ch, uint phyaddr, uint reg, uint16 v)
 		ET_ERROR(("et%d: chipphywr: did not complete\n", ch->etc->unit));
 	}
 #endif /* (defined(CONFIG_MACH_NS) || defined(CONFIG_MACH_NSP)) */
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
     uint32 addr, int_bus, bank, flags;
 
     addr = phyaddr&0xf;
@@ -2371,7 +2371,7 @@ chipphyreset(ch_t *ch, uint phyaddr)
 		ET_ERROR(("et%d: chipphyreset: reset not complete\n", ch->etc->unit));
 	}
 #endif /* (defined(CONFIG_MACH_NS) || defined(CONFIG_MACH_NSP)) */
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
     chipphywr(ch, phyaddr, 0, CTL_RESET);
     OSL_DELAY(100);
     if (chipphyrd(ch, phyaddr, 0) & CTL_RESET) {
@@ -2397,7 +2397,7 @@ chipphyinit(ch_t *ch, uint phyaddr)
 		return;
 
 	ET_TRACE(("et%d: chipphyinit: phyaddr %d\n", ch->etc->unit, phyaddr));
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 	phy5461_init(ch->etc->unit, phyaddr);
 	serdes_init(ch->etc->unit, phyaddr);
 #elif defined(CONFIG_MACH_HR2)
@@ -2497,7 +2497,7 @@ chipphyadvertise(ch_t *ch, uint phyaddr)
 static void
 chipphyenable(ch_t *ch, uint eth_num, uint phyaddr, int enable)
 {
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 	phy5461_enable_set(eth_num, phyaddr, enable);
 #endif /* (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2)) */
 #if defined(CONFIG_MACH_HR2)
diff --git a/drivers/bcmdrivers/gmac/src/include/bcmdevs.h b/drivers/bcmdrivers/gmac/src/include/bcmdevs.h
index 66283c6..7adc023 100755
--- a/drivers/bcmdrivers/gmac/src/include/bcmdevs.h
+++ b/drivers/bcmdrivers/gmac/src/include/bcmdevs.h
@@ -285,7 +285,7 @@
 
 #if defined(CONFIG_MACH_NS)
 #define BCMIPROC_CHIP_ID	BCM53010_CHIP_ID
-#elif defined(CONFIG_MACH_HX4)
+#elif (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_DNI_3048UP))
 #define BCMIPROC_CHIP_ID	BCM56340_CHIP_ID
 #elif defined(CONFIG_MACH_HR2)
 #define BCMIPROC_CHIP_ID	BCM56150_CHIP_ID
diff --git a/drivers/bcmdrivers/gmac/src/include/typedefs.h b/drivers/bcmdrivers/gmac/src/include/typedefs.h
old mode 100644
new mode 100755
index 4c787aa..1ec3ccd
--- a/drivers/bcmdrivers/gmac/src/include/typedefs.h
+++ b/drivers/bcmdrivers/gmac/src/include/typedefs.h
@@ -219,7 +219,7 @@ typedef unsigned __int64 uint64;
 
 #endif
 
-#if defined(CONFIG_MACH_NSP) || defined(CONFIG_MACH_HX4)    /* JIRA:LINUXDEV- */
+#if defined(CONFIG_MACH_NSP) || defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_DNI_3048UP)    /* JIRA:LINUXDEV- */
 #ifdef CONFIG_CPU_BIG_ENDIAN
 #define IL_BIGENDIAN
 #endif /* CONFIG_CPU_BIG_ENDIAN */
diff --git a/drivers/bcmdrivers/gmac/src/shared/aiutils.c b/drivers/bcmdrivers/gmac/src/shared/aiutils.c
index d53ade3..a9729a9 100755
--- a/drivers/bcmdrivers/gmac/src/shared/aiutils.c
+++ b/drivers/bcmdrivers/gmac/src/shared/aiutils.c
@@ -29,7 +29,7 @@
 #include <pcicfg.h>
 
 #include "siutils_priv.h"
-#if defined(CONFIG_MACH_HX4)
+#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_DNI_3048UP)
 #include "hx4_erom.h"
 #elif defined(CONFIG_MACH_HR2)
 #include "hr2_erom.h"
@@ -61,7 +61,7 @@ get_erom_ent(si_t *sih, uint32 **eromptr, uint32 mask, uint32 match)
 	while (TRUE) {
 #if defined(CONFIG_MACH_NS)
 		ent = R_REG(si_osh(sih), *eromptr);
-#elif defined(CONFIG_MACH_HX4)
+#elif defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_DNI_3048UP)
 		ent = **eromptr;
 #elif defined(CONFIG_MACH_HR2)
 		ent = **eromptr;
@@ -255,7 +255,7 @@ BCMATTACHFN(ai_scan)(si_t *sih, void *regs, uint devid)
 	case SI_BUS:
 #if defined(CONFIG_MACH_NS)
 		eromptr = (uint32 *)REG_MAP(erombase, SI_CORE_SIZE);
-#elif defined(CONFIG_MACH_HX4)
+#elif defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_DNI_3048UP)
 		eromptr = hx4_erom;
 #elif defined(CONFIG_MACH_HR2)
 		eromptr = hr2_erom;
diff --git a/drivers/bcmdrivers/gmac/src/shared/nvramstubs.c b/drivers/bcmdrivers/gmac/src/shared/nvramstubs.c
index aa8c3bd..57ca2c7 100755
--- a/drivers/bcmdrivers/gmac/src/shared/nvramstubs.c
+++ b/drivers/bcmdrivers/gmac/src/shared/nvramstubs.c
@@ -18,6 +18,8 @@
  * $Id: nvramstubs.c 325991 2012-04-05 10:16:42Z $
  */
 
+#define FAKE_NVRAM	1
+
 #include <bcm_cfg.h>
 #include <typedefs.h>
 #include <bcmutils.h>
@@ -62,7 +64,8 @@ static nvram_t fake_nvram[] = {
 	{"sdram_config",	"0x103"},
 	{"swgmacet",		"et2"},
 	{"brcmtag",			"1"},
-	//{"ethaddr",			"00:90:4c:06:a5:72"},
+	{"ethtool",			"1"},
+	{"ethaddr",			"00:90:4c:06:a5:72"},
 #ifdef FOUR_PORT_CONFIG
 	{"vlan1hwname",		"et2"},
 	{"vlan1ports",		"0 1 2 8*"},
@@ -100,6 +103,8 @@ static nvram_t fake_nvram[] = {
 #define CONFIG_NAND_BASE		0x1c000000
 #if (defined(CONFIG_MACH_NS) || defined(CONFIG_MACH_NSP))
 #define CONFIG_ENV_OFFSET		0xa0000		/* 30000-b0000 - use last 10000 for env */
+#elif defined(CONFIG_MACH_DNI_3048UP)
+#define CONFIG_ENV_OFFSET		0x100000		/* f0000-fffff - use 10000 for env */
 #else
 #define CONFIG_ENV_OFFSET		0xc0000		/* 30000-b0000 - use last 10000 for env */
 #endif
@@ -240,7 +245,7 @@ nvram_env_gmac_name(int gmac, char *name)
 	case 3:
 		sprintf(name, "eth%daddr", gmac);
 		break;
-#elif (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#elif (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 	case 0:
 		strcpy(name, "ethaddr");
 		break;
diff --git a/drivers/bcmdrivers/gpio/gpiolib.c b/drivers/bcmdrivers/gpio/gpiolib.c
old mode 100644
new mode 100755
index e7a1a66..1fe561d
--- a/drivers/bcmdrivers/gpio/gpiolib.c
+++ b/drivers/bcmdrivers/gpio/gpiolib.c
@@ -114,7 +114,7 @@ struct iproc_gpio_chip iproc_gpios_config[] = {
     },
 };
 /* CONFIG_MACH_NS */
-#elif defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2)
+#elif defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP)
 /*
  * Chip level GPIO 0-3 from CMICD,
  * GPIO 4-11 from ChipcommonA gpio pin 0 - 7
diff --git a/drivers/bcmdrivers/pcie/pcie_iproc.c b/drivers/bcmdrivers/pcie/pcie_iproc.c
index 3e0f520..043a4eb 100755
--- a/drivers/bcmdrivers/pcie/pcie_iproc.c
+++ b/drivers/bcmdrivers/pcie/pcie_iproc.c
@@ -221,7 +221,7 @@ static struct soc_pcie_port {
 	{
 	.regs_res = & soc_pcie_regs[0],
 	.owin_res = & soc_pcie_owin[0],
-#if defined(CONFIG_MACH_HX4)
+#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_DNI_3048UP)
 	.irqs = {214, 215, 216, 217, 218, 219},
 #elif defined(CONFIG_MACH_HR2)
 	.irqs = {214, 215, 216, 217, 218, 219},
diff --git a/drivers/bcmdrivers/smbus/iproc_smbus.c b/drivers/bcmdrivers/smbus/iproc_smbus.c
old mode 100644
new mode 100755
index 8ee9bf6..d4bc173
--- a/drivers/bcmdrivers/smbus/iproc_smbus.c
+++ b/drivers/bcmdrivers/smbus/iproc_smbus.c
@@ -1745,6 +1745,15 @@ static int __devinit iproc_smb_probe(struct platform_device *pdev)
 
     platform_set_drvdata(pdev, dev);
 
+    adap = &dev->adapter;
+    i2c_set_adapdata(adap, dev); /* Verify if this place is OK */
+    adap->owner = THIS_MODULE;
+    adap->class = UINT_MAX; /* Can be used by any I2C device */
+    snprintf(adap->name, sizeof(adap->name), "iproc-smb%d", pdev->id);
+    adap->algo = &iproc_smb_algorithm;
+    adap->dev.parent = &pdev->dev; /* */
+    adap->nr = pdev->id; /* We can hard code this to '0' */
+
     /* Init internal regs, disable intrs (and then clear intrs), set fifo
      * thresholds, etc.
      */
@@ -1765,15 +1774,6 @@ static int __devinit iproc_smb_probe(struct platform_device *pdev)
     printk(KERN_DEBUG "\nrequest_irq succeeded\n");
 #endif /* IPROC_SMB_DBG */
 
-    adap = &dev->adapter;
-    i2c_set_adapdata(adap, dev); /* Verify if this place is OK */
-    adap->owner = THIS_MODULE;
-    adap->class = UINT_MAX; /* Can be used by any I2C device */
-    snprintf(adap->name, sizeof(adap->name), "iproc-smb%d", pdev->id);
-    adap->algo = &iproc_smb_algorithm;
-    adap->dev.parent = &pdev->dev; /* */
-    adap->nr = pdev->id; /* We can hard code this to '0' */
-
     /*
     * I2C device drivers may be active on return from
     * i2c_add_numbered_adapter()
diff --git a/drivers/bcmdrivers/usb2h/Kconfig b/drivers/bcmdrivers/usb2h/Kconfig
old mode 100644
new mode 100755
index 8f5f8bb..0a3b2be
--- a/drivers/bcmdrivers/usb2h/Kconfig
+++ b/drivers/bcmdrivers/usb2h/Kconfig
@@ -1,7 +1,6 @@
 config IPROC_USB2H
 	tristate "USB 2.0 Host support"
 	select USB_EHCI_BCM
-	select USB_OHCI_BCM
 ## this is for core/message.c:     err = utf16s_to_utf8s
 	depends on ARCH_IPROC
 	default n
diff --git a/drivers/bcmdrivers/usb2h/bcm-iproc.c b/drivers/bcmdrivers/usb2h/bcm-iproc.c
index d006aad..015746e 100755
--- a/drivers/bcmdrivers/usb2h/bcm-iproc.c
+++ b/drivers/bcmdrivers/usb2h/bcm-iproc.c
@@ -27,6 +27,7 @@
 #include <linux/suspend.h>
 #include <mach/memory.h>
 #include <linux/version.h>
+#include <linux/i2c.h>
 
 #include "usbh_cfg.h"
 
@@ -48,22 +49,27 @@
 #define IPROC_IDM_USB2_RESET_CONTROL	 (0x18115800)
 #define IPROC_IDM_USB2_RESET_CONTROL_VA   HW_IO_PHYS_TO_VIRT(IPROC_IDM_USB2_RESET_CONTROL)
 
-#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_NSP) || defined(CONFIG_MACH_KT2)
+#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_NSP) || \
+    defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP)
 #define IPROC_IDM_USB2_IO_CONTROL_DIRECT	USB2_IDM_IDM_IO_CONTROL_DIRECT
 #define IPROC_IDM_USB2_IO_CONTROL_DIRECT_VA	HW_IO_PHYS_TO_VIRT(IPROC_IDM_USB2_IO_CONTROL_DIRECT)
 #endif
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 #define IPROC_XGPLL							0x1803fc2c
 #define IPROC_XGPLL_VA						HW_IO_PHYS_TO_VIRT(IPROC_XGPLL)
 #define IPROC_USB_PHY_CTRL					IPROC_WRAP_USBPHY_CTRL
 #define IPROC_USB_PHY_CTRL_VA				HW_IO_PHYS_TO_VIRT(IPROC_USB_PHY_CTRL)
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_DNI_3048UP))
+#define IPROC_WRAP_MISC_STATUS				0x1803fc38
+#else
 #define IPROC_WRAP_MISC_STATUS				0x1803fc28
+#endif
 #define IPROC_WRAP_MISC_STATUS_VA			HW_IO_PHYS_TO_VIRT(IPROC_WRAP_MISC_STATUS)
 #define IPROC_CLK_NDIV_40					0x80
 #define IPROC_CLK_NDIV_20					0x8C
-#define USB_CLK_NDIV_MASK					0xFE7FFE00
-#define USB_CLK_PLL_RESET_MASK				0xFF7FFE00
-#define USB_CLK_PHY_RESET_MASK				0xFFFFFE00
+#define USB_CLK_NDIV_MASK					0xFFFFFC00
+#define USB_CLK_PLL_RESET_MASK				0xFEFFFFFF
+#define USB_CLK_PHY_RESET_MASK				0xFF7FFFFF
 #define USB_CLK_NDIV_40						0x30
 #define USB_CLK_NDIV_20						0x60
 #define ChipcommonA_GPIOIn_VA				HW_IO_PHYS_TO_VIRT(ChipcommonA_GPIOInput)
@@ -276,77 +282,96 @@ static struct platform_driver usbh_driver =
 static int __init usbh_init(void)
 {
 	int usb2_reset_state;
+	uint count = 0;
 
-
-#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2))
+#if (defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_KT2) || defined(CONFIG_MACH_DNI_3048UP))
 	int clk_enable, k;
-#if defined(CONFIG_MACH_HX4)
+#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_DNI_3048UP)
 	unsigned int iClk;
 #endif
 	unsigned int USBClk, usbdgpiopwr, pllStatus;
 
+#if defined(CONFIG_MACH_DNI_3048UP)
+	struct i2c_adapter *i2c_adap;
+	unsigned char data = 0;
+
+	i2c_adap = i2c_get_adapter(0);
+	i2c_smbus_xfer(i2c_adap, I2C_CPLD_ADDR, 0, I2C_SMBUS_READ, CPLD_SW_RESET_REG, I2C_SMBUS_BYTE_DATA, &data);
+	/*printk("data1 = 0x%x \n",data);*/
+	data &= ~(CPLD_RST_OOB);
+	i2c_smbus_xfer(i2c_adap, I2C_CPLD_ADDR, 0, I2C_SMBUS_WRITE, CPLD_SW_RESET_REG, I2C_SMBUS_BYTE_DATA, &data);
+#endif
+
+#if defined(CONFIG_MACH_HX4)
 	/* turn off power for USB device connected to the host */
 	usbdgpiopwr = readl_relaxed(ChipcommonA_GPIOOut_VA);
 	usbdgpiopwr |= 0x2;
 	writel_relaxed(usbdgpiopwr, ChipcommonA_GPIOOut_VA);
 	writel_relaxed(0x2, ChipcommonA_GPIOOutEn_VA);
+#endif
 
 	/* Do USB PHY reset */
 	mdelay(100);
 	USBClk = readl_relaxed(IPROC_USB_PHY_CTRL_VA);
-	/* bring phy pll out of reset if not done already */
-	if ((USBClk & 0x01000000) == 0 )
-	{
-		USBClk |= 0x01000000;
-		writel_relaxed(USBClk, IPROC_USB_PHY_CTRL_VA);
-		pllStatus = readl_relaxed(IPROC_WRAP_MISC_STATUS_VA);
-		for (k = 0; k < 100000; k++)
-		{
-			if ((pllStatus & 2) == 2)
-			{
-				printk("USB phy pll locked\n");
-				break;
-			}
-			pllStatus = readl_relaxed(IPROC_WRAP_MISC_STATUS_VA);
-		}
-	}
-	writel_relaxed(USBClk & (~(1<<23)), IPROC_USB_PHY_CTRL_VA);
+	printk("iClk = %08x, USBClk = %08x\n", iClk, USBClk);
+	USBClk &= USB_CLK_PHY_RESET_MASK;
+	writel_relaxed(USBClk, IPROC_USB_PHY_CTRL_VA);
+	mdelay(100);
+	
+	USBClk &= USB_CLK_PLL_RESET_MASK;
+	writel_relaxed(USBClk, IPROC_USB_PHY_CTRL_VA);
+	udelay(10);
+	
 	clk_enable = readl_relaxed(IPROC_IDM_USB2_IO_CONTROL_DIRECT_VA);
 	printk("Initial usb2h clock is: %08x\n", clk_enable);
 	clk_enable |= 1;
 	writel_relaxed(clk_enable, IPROC_IDM_USB2_IO_CONTROL_DIRECT_VA);
 	clk_enable = readl_relaxed(IPROC_IDM_USB2_IO_CONTROL_DIRECT_VA);
 	printk("Initial usb2h clock now is: %08x\n", clk_enable);
-#if defined(CONFIG_MACH_HX4)
+#if defined(CONFIG_MACH_HX4) || defined(CONFIG_MACH_DNI_3048UP)
 	iClk = readl_relaxed(IPROC_XGPLL_VA);
 	USBClk = readl_relaxed(IPROC_USB_PHY_CTRL_VA);
-	printk("iClk = %08x, USBClk = %08x\n", iClk, USBClk);
+	/*printk("iClk = %08x, USBClk = %08x\n", iClk, USBClk);*/
 	if ((iClk & 0xff) == IPROC_CLK_NDIV_40)
 	{
 		writel_relaxed((USBClk & USB_CLK_NDIV_MASK) | USB_CLK_NDIV_40, IPROC_USB_PHY_CTRL_VA);
 		udelay(10);
-		writel_relaxed((USBClk & USB_CLK_PLL_RESET_MASK) | USB_CLK_NDIV_40, IPROC_USB_PHY_CTRL_VA);
-		udelay(10);
-		writel_relaxed((USBClk & USB_CLK_PHY_RESET_MASK) | USB_CLK_NDIV_40, IPROC_USB_PHY_CTRL_VA);
-		udelay(10);
-		USBClk = readl_relaxed(IPROC_USB_PHY_CTRL_VA);
 		printk("iClk = %08x, USBClk = %08x\n", iClk, USBClk);
 	}
 	else if ((iClk & 0xff) == IPROC_CLK_NDIV_20)
 	{
 		writel_relaxed((USBClk & USB_CLK_NDIV_MASK) | USB_CLK_NDIV_20, IPROC_USB_PHY_CTRL_VA);
 		udelay(10);
-		writel_relaxed((USBClk & USB_CLK_PLL_RESET_MASK) | USB_CLK_NDIV_20, IPROC_USB_PHY_CTRL_VA);
-		udelay(10);
-		writel_relaxed((USBClk & USB_CLK_PHY_RESET_MASK) | USB_CLK_NDIV_20, IPROC_USB_PHY_CTRL_VA);
-		udelay(10);
-		USBClk = readl_relaxed(IPROC_USB_PHY_CTRL_VA);
 		printk("iClk = %08x, USBClk = %08x\n", iClk, USBClk);
 	}
 #endif
+
+	/* bring USB PHY out of reset */
+	USBClk = readl_relaxed(IPROC_USB_PHY_CTRL_VA);
+	USBClk |= ~USB_CLK_PLL_RESET_MASK;
+	writel_relaxed(USBClk, IPROC_USB_PHY_CTRL_VA);
+	mdelay(10);
+
+	do {
+        pllStatus = readl_relaxed(IPROC_WRAP_MISC_STATUS_VA);
+        if ((pllStatus & 2) == 2) {
+			printk("USB phy pll locked\n");
+            break;
+        } else {
+            udelay(10);
+            count ++;
+        }
+    } while(count <= 10);
+    if (count > 10) {
+        printk(KERN_WARNING "%s : PLL not lock! IPROC_WRAP_MISC_STATUS = 0x%08lx\n",
+               __FUNCTION__, pllStatus);
+    }
+	
+	USBClk = readl_relaxed(IPROC_USB_PHY_CTRL_VA);
+	USBClk |= ~USB_CLK_PHY_RESET_MASK;
+	writel_relaxed(USBClk, IPROC_USB_PHY_CTRL_VA);
 	mdelay(100);
-	writel_relaxed(USBClk | (1<<23), IPROC_USB_PHY_CTRL_VA);
-	udelay(100);
+	
 #endif
 #if defined(CONFIG_MACH_NSP)
 	int clk_enable;
@@ -374,6 +399,15 @@ static int __init usbh_init(void)
 	writel_relaxed(usbdgpiopwr, ChipcommonA_GPIOOut_VA);
 	writel_relaxed(0x2, ChipcommonA_GPIOOutEn_VA);
 #endif
+
+#if defined(CONFIG_MACH_DNI_3048UP)
+	data = 0;
+	i2c_smbus_xfer(i2c_adap, I2C_CPLD_ADDR, 0, I2C_SMBUS_READ, CPLD_SW_RESET_REG, I2C_SMBUS_BYTE_DATA, &data);
+	/*printk("data2 = 0x%x \n",data);*/
+	data |= CPLD_RST_OOB;
+	i2c_smbus_xfer(i2c_adap, I2C_CPLD_ADDR, 0, I2C_SMBUS_WRITE, CPLD_SW_RESET_REG, I2C_SMBUS_BYTE_DATA, &data);
+#endif
+
 	return platform_driver_register(&usbh_driver);
 }
 
diff --git a/drivers/mtd/devices/m25p80.c b/drivers/mtd/devices/m25p80.c
old mode 100644
new mode 100755
index 3d6beb7..2287c9b
--- a/drivers/mtd/devices/m25p80.c
+++ b/drivers/mtd/devices/m25p80.c
@@ -936,11 +936,13 @@ static int __devinit m25p_probe(struct spi_device *spi)
 	if (info->addr_width)
 		flash->addr_width = info->addr_width;
 	else {
+#ifndef CONFIG_MACH_DNI_3048UP
 		/* enable 4-byte addressing if the device exceeds 16MiB */
 		if (flash->mtd.size > 0x1000000) {
 			flash->addr_width = 4;
 			set_4byte(flash, info->jedec_id, 1);
 		} else
+#endif /* !CONFIG_M25PXX_STAY_IN_3BYTE_MODE */
 			flash->addr_width = 3;
 	}
 
diff --git a/drivers/rtc/rtc-m41t80.c b/drivers/rtc/rtc-m41t80.c
old mode 100644
new mode 100755
index 64aedd8..5d14580
--- a/drivers/rtc/rtc-m41t80.c
+++ b/drivers/rtc/rtc-m41t80.c
@@ -89,26 +89,86 @@ struct m41t80_data {
 	struct rtc_device *rtc;
 };
 
+#define BLOCK_DATA_MAX_TRIES 10
+
+static s32 m41t80_read_block_data_once(const struct i2c_client *client,
+				u8 command, u8 length, u8 *values)
+{
+	s32 i, data;
+
+	for (i = 0; i < length; i++) {
+		data = i2c_smbus_read_byte_data(client, command + i);
+		if (data < 0)
+			return data;
+		values[i] = data;
+	}
+	return i;
+}
+
+static s32 m41t80_read_block_data(const struct i2c_client *client, u8 command,
+				  u8 length, u8 *values)
+{
+	u8 oldvalues[I2C_SMBUS_BLOCK_MAX];
+	s32 ret;
+	int tries = 0;
+
+	dev_dbg(&client->dev, "m41t80_read_block_data (length=%d)\n", length);
+	ret = m41t80_read_block_data_once(client, command, length, values);
+	if (ret < 0)
+		return ret;
+	do {
+		if (++tries > BLOCK_DATA_MAX_TRIES) {
+			dev_err(&client->dev,
+				"m41t80_read_block_data failed\n");
+			return -EIO;
+		}
+		memcpy(oldvalues, values, length);
+		ret = m41t80_read_block_data_once(client, command, length,
+						  values);
+		if (ret < 0)
+			return ret;
+	} while (memcmp(oldvalues, values, length));
+	return length;
+}
+
+static s32 m41t80_write_block_data(const struct i2c_client *client, u8 command,
+				   u8 length, const u8 *values)
+{
+	u8 currvalues[I2C_SMBUS_BLOCK_MAX];
+	int tries = 0;
+
+	dev_dbg(&client->dev, "m41t80_write_block_data (length=%d)\n", length);
+	do {
+		s32 i, ret;
+
+		if (++tries > BLOCK_DATA_MAX_TRIES) {
+			dev_err(&client->dev,
+				"m41t80_write_block_data failed\n");
+			return -EIO;
+		}
+		for (i = 0; i < length; i++) {
+			ret = i2c_smbus_write_byte_data(client, command + i,
+							values[i]);
+			if (ret < 0)
+				return ret;
+		}
+		ret = m41t80_read_block_data_once(client, command, length,
+						  currvalues);
+		if (ret < 0)
+			return ret;
+	} while (memcmp(currvalues, values, length));
+	return length;
+}
+
+
 static int m41t80_get_datetime(struct i2c_client *client,
 			       struct rtc_time *tm)
 {
-	u8 buf[M41T80_DATETIME_REG_SIZE], dt_addr[1] = { M41T80_REG_SEC };
-	struct i2c_msg msgs[] = {
-		{
-			.addr	= client->addr,
-			.flags	= 0,
-			.len	= 1,
-			.buf	= dt_addr,
-		},
-		{
-			.addr	= client->addr,
-			.flags	= I2C_M_RD,
-			.len	= M41T80_DATETIME_REG_SIZE - M41T80_REG_SEC,
-			.buf	= buf + M41T80_REG_SEC,
-		},
-	};
+	u8 buf[M41T80_DATETIME_REG_SIZE];
 
-	if (i2c_transfer(client->adapter, msgs, 2) < 0) {
+	if (m41t80_read_block_data (client, M41T80_REG_SEC,
+		M41T80_DATETIME_REG_SIZE - M41T80_REG_SEC,
+		buf + M41T80_REG_SEC) < 0) {
 		dev_err(&client->dev, "read error\n");
 		return -EIO;
 	}
@@ -130,32 +190,11 @@ static int m41t80_set_datetime(struct i2c_client *client, struct rtc_time *tm)
 {
 	u8 wbuf[1 + M41T80_DATETIME_REG_SIZE];
 	u8 *buf = &wbuf[1];
-	u8 dt_addr[1] = { M41T80_REG_SEC };
-	struct i2c_msg msgs_in[] = {
-		{
-			.addr	= client->addr,
-			.flags	= 0,
-			.len	= 1,
-			.buf	= dt_addr,
-		},
-		{
-			.addr	= client->addr,
-			.flags	= I2C_M_RD,
-			.len	= M41T80_DATETIME_REG_SIZE - M41T80_REG_SEC,
-			.buf	= buf + M41T80_REG_SEC,
-		},
-	};
-	struct i2c_msg msgs[] = {
-		{
-			.addr	= client->addr,
-			.flags	= 0,
-			.len	= 1 + M41T80_DATETIME_REG_SIZE,
-			.buf	= wbuf,
-		 },
-	};
 
 	/* Read current reg values into buf[1..7] */
-	if (i2c_transfer(client->adapter, msgs_in, 2) < 0) {
+	if (m41t80_read_block_data(client, M41T80_REG_SEC, 
+		M41T80_DATETIME_REG_SIZE - M41T80_REG_SEC,
+		buf + M41T80_REG_SEC) < 0) {
 		dev_err(&client->dev, "read error\n");
 		return -EIO;
 	}
@@ -178,7 +217,7 @@ static int m41t80_set_datetime(struct i2c_client *client, struct rtc_time *tm)
 	/* assume 20YY not 19YY */
 	buf[M41T80_REG_YEAR] = bin2bcd(tm->tm_year % 100);
 
-	if (i2c_transfer(client->adapter, msgs, 1) != 1) {
+	if (m41t80_write_block_data(client, 0, M41T80_DATETIME_REG_SIZE, wbuf + 1) != M41T80_DATETIME_REG_SIZE) {
 		dev_err(&client->dev, "write error\n");
 		return -EIO;
 	}
@@ -241,31 +280,8 @@ static int m41t80_rtc_set_alarm(struct device *dev, struct rtc_wkalrm *t)
 	u8 wbuf[1 + M41T80_ALARM_REG_SIZE];
 	u8 *buf = &wbuf[1];
 	u8 *reg = buf - M41T80_REG_ALARM_MON;
-	u8 dt_addr[1] = { M41T80_REG_ALARM_MON };
-	struct i2c_msg msgs_in[] = {
-		{
-			.addr	= client->addr,
-			.flags	= 0,
-			.len	= 1,
-			.buf	= dt_addr,
-		},
-		{
-			.addr	= client->addr,
-			.flags	= I2C_M_RD,
-			.len	= M41T80_ALARM_REG_SIZE,
-			.buf	= buf,
-		},
-	};
-	struct i2c_msg msgs[] = {
-		{
-			.addr	= client->addr,
-			.flags	= 0,
-			.len	= 1 + M41T80_ALARM_REG_SIZE,
-			.buf	= wbuf,
-		 },
-	};
 
-	if (i2c_transfer(client->adapter, msgs_in, 2) < 0) {
+	if (m41t80_read_block_data(client, M41T80_REG_ALARM_MON, M41T80_ALARM_REG_SIZE, buf) < 0) {
 		dev_err(&client->dev, "read error\n");
 		return -EIO;
 	}
@@ -289,7 +305,7 @@ static int m41t80_rtc_set_alarm(struct device *dev, struct rtc_wkalrm *t)
 	else
 		reg[M41T80_REG_ALARM_DAY] |= 0x40;
 
-	if (i2c_transfer(client->adapter, msgs, 1) != 1) {
+	if (m41t80_write_block_data(client, M41T80_REG_ALARM_MON, M41T80_ALARM_REG_SIZE, wbuf + 1) != 1) {
 		dev_err(&client->dev, "write error\n");
 		return -EIO;
 	}
@@ -309,24 +325,9 @@ static int m41t80_rtc_read_alarm(struct device *dev, struct rtc_wkalrm *t)
 {
 	struct i2c_client *client = to_i2c_client(dev);
 	u8 buf[M41T80_ALARM_REG_SIZE + 1]; /* all alarm regs and flags */
-	u8 dt_addr[1] = { M41T80_REG_ALARM_MON };
 	u8 *reg = buf - M41T80_REG_ALARM_MON;
-	struct i2c_msg msgs[] = {
-		{
-			.addr	= client->addr,
-			.flags	= 0,
-			.len	= 1,
-			.buf	= dt_addr,
-		},
-		{
-			.addr	= client->addr,
-			.flags	= I2C_M_RD,
-			.len	= M41T80_ALARM_REG_SIZE + 1,
-			.buf	= buf,
-		},
-	};
 
-	if (i2c_transfer(client->adapter, msgs, 2) < 0) {
+	if (m41t80_read_block_data(client, M41T80_REG_ALARM_MON, M41T80_ALARM_REG_SIZE + 1, buf) < 0) {
 		dev_err(&client->dev, "read error\n");
 		return -EIO;
 	}
@@ -509,14 +510,6 @@ static int boot_flag;
 static void wdt_ping(void)
 {
 	unsigned char i2c_data[2];
-	struct i2c_msg msgs1[1] = {
-		{
-			.addr	= save_client->addr,
-			.flags	= 0,
-			.len	= 2,
-			.buf	= i2c_data,
-		},
-	};
 	struct m41t80_data *clientdata = i2c_get_clientdata(save_client);
 
 	i2c_data[0] = 0x09;		/* watchdog register */
@@ -536,7 +529,7 @@ static void wdt_ping(void)
 	if (clientdata->features & M41T80_FEATURE_WD)
 		i2c_data[1] &= ~M41T80_WATCHDOG_RB2;
 
-	i2c_transfer(save_client->adapter, msgs1, 1);
+	m41t80_write_block_data (save_client, 0x09, 1, &i2c_data[1]);
 }
 
 /**
@@ -547,35 +540,13 @@ static void wdt_ping(void)
 static void wdt_disable(void)
 {
 	unsigned char i2c_data[2], i2c_buf[0x10];
-	struct i2c_msg msgs0[2] = {
-		{
-			.addr	= save_client->addr,
-			.flags	= 0,
-			.len	= 1,
-			.buf	= i2c_data,
-		},
-		{
-			.addr	= save_client->addr,
-			.flags	= I2C_M_RD,
-			.len	= 1,
-			.buf	= i2c_buf,
-		},
-	};
-	struct i2c_msg msgs1[1] = {
-		{
-			.addr	= save_client->addr,
-			.flags	= 0,
-			.len	= 2,
-			.buf	= i2c_data,
-		},
-	};
 
 	i2c_data[0] = 0x09;
-	i2c_transfer(save_client->adapter, msgs0, 2);
+	m41t80_read_block_data(save_client, 0x09, 1, i2c_buf);
 
 	i2c_data[0] = 0x09;
 	i2c_data[1] = 0x00;
-	i2c_transfer(save_client->adapter, msgs1, 1);
+	m41t80_write_block_data(save_client, 0x09, 1, &i2c_data[1]);
 }
 
 /**
@@ -777,8 +748,7 @@ static int m41t80_probe(struct i2c_client *client,
 	struct rtc_time tm;
 	struct m41t80_data *clientdata = NULL;
 
-	if (!i2c_check_functionality(client->adapter, I2C_FUNC_I2C
-				     | I2C_FUNC_SMBUS_BYTE_DATA)) {
+	if (!i2c_check_functionality(client->adapter, I2C_FUNC_SMBUS_BYTE_DATA)) {
 		rc = -ENODEV;
 		goto exit;
 	}
diff --git a/include/linux/i2c.h b/include/linux/i2c.h
old mode 100644
new mode 100755
index 07d103a..5b1fd32
--- a/include/linux/i2c.h
+++ b/include/linux/i2c.h
@@ -602,4 +602,10 @@ union i2c_smbus_data {
 #define I2C_SMBUS_BLOCK_PROC_CALL   7		/* SMBus 2.0 */
 #define I2C_SMBUS_I2C_BLOCK_DATA    8
 
+#if defined(CONFIG_MACH_DNI_3048UP)
+#define I2C_CPLD_ADDR	0x28
+#define CPLD_SW_RESET_REG 0x0e
+#define CPLD_RST_OOB 0x40
+#endif /* CONFIG_MACH_DNI_3048UP */
+
 #endif /* _LINUX_I2C_H */
