diff --git a/drivers/i2c/busses/i2c-ismt.c b/drivers/i2c/busses/i2c-ismt.c
old mode 100644
new mode 100755
index 09a3e03..a82c962
--- a/drivers/i2c/busses/i2c-ismt.c
+++ b/drivers/i2c/busses/i2c-ismt.c
@@ -192,7 +192,11 @@ static DEFINE_PCI_DEVICE_TABLE(ismt_ids) = {
 MODULE_DEVICE_TABLE(pci, ismt_ids);
 
 /* Bus speed control bits for slow debuggers - refer to the docs for usage */
+#if 1 /* Foxconn: Set i2c bus_speed to 100kHZ */
+static unsigned int bus_speed = 100;
+#else
 static unsigned int bus_speed;
+#endif
 module_param(bus_speed, uint, S_IRUGO);
 MODULE_PARM_DESC(bus_speed, "Bus Speed in kHz (0 = BIOS default)");
 
