diff --git a/include/mb_detect.h b/include/mb_detect.h
new file mode 100644
index 0000000..841f2db
--- /dev/null
+++ b/include/mb_detect.h
@@ -0,0 +1,70 @@
+#ifndef __MB_DETECT_H
+#define __MB_DETECT_H
+
+typedef __u8 u8;
+typedef __u32 u32;
+typedef __s32 s32;
+
+#define MB_DETECT_I2C_BUS 1
+
+#define MAX_NUM_MB 0x3f
+enum {
+    MB_HURACAN_BMC = 0,
+    MB_HURACAN,
+    MB_CABRERAIII_BMC,
+    MB_CABRERAIII,
+    MB_SESTO_BMC,
+    MB_SESTO,
+    MB_NCIIX_BMC,
+    MB_NCIIX,
+    MB_ASTERION_BMC,
+    MB_ASTERION,
+    MB_HURACAN_A_BMC,
+    MB_HURACAN_A,
+    MB_EVORA,
+    MB_EUROPA,
+    MB_LAST
+};
+
+#define SYS_PRODUCT_NAME_SIZE           64  /* max. product name size */
+#define SYS_MODEL_NAME_SIZE             64  /* max. model name size */
+typedef struct {
+    u8 motherboard_id;
+    u8 projectName[SYS_PRODUCT_NAME_SIZE+1];
+    u8 productName[SYS_PRODUCT_NAME_SIZE+1];
+    u8 modelName[SYS_MODEL_NAME_SIZE+1];
+} SysBoardInfo;
+
+typedef struct {
+    u8 i2c_mux;
+    u8 i2c_channel;
+    u8 i2c_dev;
+    u8 i2c_bus[32];
+} SysEEPROMPath;
+
+extern u8 mb_id_get(void);
+extern u8 mb_index_get(void);
+extern s32 sys_eeprom_path_get(SysEEPROMPath *ep);
+extern s32 sys_board_info_get(SysBoardInfo *bdinfo);
+
+/* Model ID Definition */
+typedef enum
+{
+    HURACAN_WITH_BMC = 0x0,
+    HURACAN_WITHOUT_BMC,
+    CABRERAIII_WITH_BMC,
+    CABRERAIII_WITHOUT_BMC,
+    SESTO_WITH_BMC,
+    SESTO_WITHOUT_BMC,
+    NCIIX_WITH_BMC,
+    NCIIX_WITHOUT_BMC,
+    ASTERION_WITH_BMC,
+    ASTERION_WITHOUT_BMC,
+    HURACAN_A_WITH_BMC,
+    HURACAN_A_WITHOUT_BMC,
+    EVORA_WITHOUT_BMC = 0xE0,
+    EUROPA_WITHOUT_BMC = 0xE1,
+    MODEL_ID_LAST
+} modelId_t;
+
+#endif  /* __MB_DETECT_H */
