From 367d466ce31a3e67129cbb66cbe5352c73e11091 Mon Sep 17 00:00:00 2001
From: Shaohui Xie <Shaohui.Xie@freescale.com>
Date: Thu, 24 Sep 2015 18:20:32 +0800
Subject: [PATCH 28/74] armv8: ls2085ardb: enable CONFIG_PHY_AQUANTIA

To support on board Aquantia's PHY AQR405.

Signed-off-by: Shaohui Xie <Shaohui.Xie@freescale.com>
Reviewed-by: York Sun <yorksun@freescale.com>
---
 include/configs/ls2085ardb.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/include/configs/ls2085ardb.h b/include/configs/ls2085ardb.h
index 10482a9..5188dea 100644
--- a/include/configs/ls2085ardb.h
+++ b/include/configs/ls2085ardb.h
@@ -318,6 +318,7 @@ unsigned long get_board_sys_clk(void);
 /* MAC/PHY configuration */
 #ifdef CONFIG_FSL_MC_ENET
 #define CONFIG_PHYLIB_10G
+#define CONFIG_PHY_AQUANTIA
 #define CONFIG_PHY_CORTINA
 #define CONFIG_PHYLIB
 #define	CONFIG_SYS_CORTINA_FW_IN_NOR
-- 
2.1.0.27.g96db324

