From 42909e014685206626a7db9190813650f0367ad4 Mon Sep 17 00:00:00 2001
From: Rai Harninder-B01044 <harninder.rai@freescale.com>
Date: Tue, 27 Oct 2015 17:19:11 +0530
Subject: [PATCH 018/124] pci/layerscape: Add compatible string for LS2080

Signed-off-by: Rai Harninder-B01044 <harninder.rai@freescale.com>
---
 drivers/pci/host/pci-layerscape.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/pci/host/pci-layerscape.c b/drivers/pci/host/pci-layerscape.c
index ece1337..262b32e 100644
--- a/drivers/pci/host/pci-layerscape.c
+++ b/drivers/pci/host/pci-layerscape.c
@@ -164,6 +164,7 @@ static struct dw_host_ops ls2_dw_host_ops = {
 static const struct of_device_id ls_pcie_of_match[] = {
 	{ .compatible = "fsl,ls1021a-pcie", .data = &ls1_dw_host_ops },
 	{ .compatible = "fsl,ls2085a-pcie", .data = &ls2_dw_host_ops },
+	{ .compatible = "fsl,ls2080a-pcie", .data = &ls2_dw_host_ops },
 	{ },
 };
 MODULE_DEVICE_TABLE(of, ls_pcie_of_match);
-- 
2.1.0.27.g96db324

