From 444fe6c5c0fee1ffcae03a3f487d593afc2e6eb2 Mon Sep 17 00:00:00 2001
From: Mingkai Hu <mingkai.hu@nxp.com>
Date: Wed, 22 Jun 2016 13:09:27 +0800
Subject: [PATCH 034/124] layerscape/pci: fix linkup issue

Signed-off-by: Mingkai Hu <mingkai.hu@nxp.com>
---
 drivers/pci/host/pci-layerscape.c | 13 +++++++++----
 1 file changed, 9 insertions(+), 4 deletions(-)

diff --git a/drivers/pci/host/pci-layerscape.c b/drivers/pci/host/pci-layerscape.c
index 8b6a59e..baa1232 100644
--- a/drivers/pci/host/pci-layerscape.c
+++ b/drivers/pci/host/pci-layerscape.c
@@ -329,11 +329,16 @@ static void ls1021_pcie_host_init(struct pcie_port *pp)
 static int ls_pcie_link_up(struct pcie_port *pp)
 {
 	struct ls_pcie *pcie = to_ls_pcie(pp);
-	u32 state;
+	u32 state, offset;
 
-	state = (ioread32(pcie->lut + PCIE_LUT_DBG) >>
-		 pcie->drvdata->ltssm_shift) &
-		 LTSSM_STATE_MASK;
+	if (of_get_property(pp->dev->of_node, "fsl,lut_diff", NULL))
+		offset = 0x407fc;
+	else
+		offset = PCIE_LUT_DBG;
+
+	state = (ioread32(pcie->lut + offset) >>
+			pcie->drvdata->ltssm_shift) &
+		LTSSM_STATE_MASK;
 
 	if (state < LTSSM_PCIE_L0)
 		return 0;
-- 
2.1.0.27.g96db324

