From 5d023205e8111f394d5d95f4a014a43b32a2cbd9 Mon Sep 17 00:00:00 2001
From: Diana Craciun <Diana.Craciun@freescale.com>
Date: Tue, 7 Jul 2015 17:19:48 +0300
Subject: [PATCH 042/124] Added MSI support in host generic driver

Signed-off-by: Diana Craciun <Diana.Craciun@freescale.com>
---
 drivers/pci/host/pci-host-generic.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/drivers/pci/host/pci-host-generic.c b/drivers/pci/host/pci-host-generic.c
index 265dd25..46f046b 100644
--- a/drivers/pci/host/pci-host-generic.c
+++ b/drivers/pci/host/pci-host-generic.c
@@ -205,6 +205,18 @@ static int gen_pci_parse_map_cfg_windows(struct gen_pci *pci)
 	return 0;
 }
 
+static void gen_pci_enable_msi(struct gen_pci *pci, struct pci_bus *bus)
+{
+	struct device_node *msi_node;
+	struct device *dev = pci->host.dev.parent;
+	struct device_node *np = dev->of_node;
+
+	/*setup MSIs */
+	msi_node = of_parse_phandle(np, "msi-parent", 0);
+	if (msi_node)
+		bus->msi = of_pci_find_msi_chip_by_node(msi_node);
+}
+
 static int gen_pci_probe(struct platform_device *pdev)
 {
 	int err;
@@ -262,6 +274,8 @@ static int gen_pci_probe(struct platform_device *pdev)
 		return -ENODEV;
 	}
 
+	gen_pci_enable_msi(pci, bus);
+
 	pci_fixup_irqs(pci_common_swizzle, of_irq_parse_and_map_pci);
 
 	if (!pci_has_flag(PCI_PROBE_ONLY)) {
-- 
2.1.0.27.g96db324

