From 7d15d2688bd47cfdb6e0174e7a492f2d6ca9d77b Mon Sep 17 00:00:00 2001
From: Zhao Qiang <qiang.zhao@nxp.com>
Date: Fri, 1 Jun 2018 09:34:56 +0800
Subject: [PATCH 328/448] smmu: fix compile issue

Signed-off-by: Zhao Qiang <qiang.zhao@nxp.com>
---
 drivers/iommu/arm-smmu.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/iommu/arm-smmu.c b/drivers/iommu/arm-smmu.c
index 382a1d3..342319a 100644
--- a/drivers/iommu/arm-smmu.c
+++ b/drivers/iommu/arm-smmu.c
@@ -49,7 +49,7 @@
 #include <linux/spinlock.h>
 
 #include <linux/amba/bus.h>
-#include "../staging/fsl-mc/include/mc-bus.h"
+#include <linux/fsl/mc.h>
 
 #include "io-pgtable.h"
 
-- 
1.7.1

