From 7a1ebfcfafcce7c846a0efe483309494f8052f69 Mon Sep 17 00:00:00 2001
From: Zhao Qiang <qiang.zhao@nxp.com>
Date: Fri, 8 Jun 2018 17:18:03 +0800
Subject: [PATCH 347/448] crypto: talitos: fix compile issue

Signed-off-by: Zhao Qiang <qiang.zhao@nxp.com>
---
 drivers/crypto/talitos.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/drivers/crypto/talitos.c b/drivers/crypto/talitos.c
index 614713f..c905b34 100644
--- a/drivers/crypto/talitos.c
+++ b/drivers/crypto/talitos.c
@@ -1244,7 +1244,8 @@ static int ipsec_esp(struct talitos_edesc *edesc, struct aead_request *areq,
 	 * In case of SEC 2.x+, cipher in len must include only the ciphertext,
 	 * while extent is used for ICV len.
 	 */
-	if (is_ipsec_esp && (desc->hdr & DESC_HDR_MODE1_MDEU_CICV))
+	if ((edesc->desc.hdr & DESC_HDR_TYPE_IPSEC_ESP) &&
+	    (desc->hdr & DESC_HDR_MODE1_MDEU_CICV))
 		desc->ptr[4].len = cpu_to_be16(cryptlen);
 
 	if (ret > 1) {
-- 
1.7.1

