From 17c2ed1f7aa8a15a9613c80446e0b5b478237696 Mon Sep 17 00:00:00 2001
From: Camelia Groza <camelia.groza@nxp.com>
Date: Mon, 26 Mar 2018 17:12:17 +0300
Subject: [PATCH 377/448] fsl_qbman: ceetm: export the qman_ceetm_query_cq call

Signed-off-by: Camelia Groza <camelia.groza@nxp.com>
---
 include/linux/fsl_qman.h |   12 ++++++++++++
 1 files changed, 12 insertions(+), 0 deletions(-)

diff --git a/include/linux/fsl_qman.h b/include/linux/fsl_qman.h
index 4e4b21d..56c05fd 100644
--- a/include/linux/fsl_qman.h
+++ b/include/linux/fsl_qman.h
@@ -3813,6 +3813,18 @@ int qman_ceetm_query_lfqmt(int lfqid,
 			   struct qm_mcr_ceetm_lfqmt_query *lfqmt_query);
 
 /**
+ * qman_ceetm_query_cq - Queries a CEETM CQ
+ * @cqid: the channel ID (first byte) followed by the CQ idx
+ * @dcpid: CEETM portal ID
+ * @cq_query: storage for the queried CQ fields
+ *
+ * Returns zero for success or -EIO if the query command returns error.
+ *
+*/
+int qman_ceetm_query_cq(unsigned int cqid, unsigned int dcpid,
+			struct qm_mcr_ceetm_cq_query *cq_query);
+
+/**
  * qman_ceetm_query_write_statistics - Query (and optionally write) statistics
  * @cid: Target ID (CQID or CCGRID)
  * @dcp_idx: CEETM portal ID
-- 
1.7.1

