diff --git a/sys_eeprom/24cXX.c b/sys_eeprom/24cXX.c
index be194ad..ea63ba1
--- a/sys_eeprom/24cXX.c
+++ b/sys_eeprom/24cXX.c
@@ -33,7 +33,7 @@ static int i2c_write_1b(struct eeprom *e, __u8 buf)
 	r = i2c_smbus_write_byte(e->fd, buf);
 	if(r < 0)
 		fprintf(stderr, "Error i2c_write_1b: %s\n", strerror(errno));
-	usleep(10);
+	usleep(10*1000);
 	return r;
 }
 
@@ -44,7 +44,7 @@ static int i2c_write_2b(struct eeprom *e, __u8 buf[2])
 	r = i2c_smbus_write_byte_data(e->fd, buf[0], buf[1]);
 	if(r < 0)
 		fprintf(stderr, "Error i2c_write_2b: %s\n", strerror(errno));
-	usleep(1000);
+	usleep(10*1000);
 	return r;
 }
 
@@ -56,7 +56,7 @@ static int i2c_write_3b(struct eeprom *e, __u8 buf[3])
 	r = i2c_smbus_write_word_data(e->fd, buf[0], buf[2] << 8 | buf[1]);
 	if(r < 0)
 		fprintf(stderr, "Error i2c_write_3b: %s\n", strerror(errno));
-	usleep(10);
+	usleep(10*1000);
 	return r;
 }
 
