From d0fcaf237405fa2d1576b8455b30c3bb680f64d7 Mon Sep 17 00:00:00 2001
From: Luffy <luffy.cheng@quantatw.com>
Date: Fri, 13 Mar 2015 09:53:53 +0800
Subject: Correct clk_adjust value for MPC8541

---
 board/quanta/quanta_lb9/ddr.c |   11 ++++++++---
 1 files changed, 8 insertions(+), 3 deletions(-)

diff --git a/board/quanta/quanta_lb9/ddr.c b/board/quanta/quanta_lb9/ddr.c
index 78d73b0..885be6e 100644
--- a/board/quanta/quanta_lb9/ddr.c
+++ b/board/quanta/quanta_lb9/ddr.c
@@ -23,10 +23,15 @@ void fsl_ddr_board_options(memctl_options_t *popts,
 	 *	- ???
 	 *
 	 * This needs to be determined on a board-by-board basis.
-	 *	0110	3/4 cycle late
-	 *	0111	7/8 cycle late
+	 *
+	 *  MPC8541 / MPC8555 specific
+	 *	000		 no cycle late
+	 *	001		1/4 cycle late
+	 *	010		1/2 cycle late
+	 *	011		3/4 cycle late
+	 *	100		  1 cycle late
 	 */
-	popts->clk_adjust = 6;
+	popts->clk_adjust = 3;
 
 	/*
 	 * Factors to consider for CPO:
-- 
1.7.3.4

