diff --git a/drivers/base/dd.c b/drivers/base/dd.c
old mode 100644
new mode 100644
index 142e3d6..f2e6a3f
--- a/drivers/base/dd.c
+++ b/drivers/base/dd.c
@@ -143,10 +143,15 @@ probe_failed:
 	dev->driver = NULL;
 
 	if (ret != -ENODEV && ret != -ENXIO) {
-		/* driver matched but the probe failed */
-		printk(KERN_WARNING
-		       "%s: probe of %s failed with error %d\n",
-		       drv->name, dev_name(dev), ret);
+		/* Skip the probe error message for these three devices on Alphanetworks platforms. */
+		if (strncmp(dev_name(dev), "0000:00:14.0", 12) &&
+			strncmp(dev_name(dev), "0000:00:14.2", 12) &&
+			strncmp(dev_name(dev), "0000:00:14.3", 12)) {
+			/* driver matched but the probe failed */
+			printk(KERN_WARNING
+			       "%s: probe of %s failed with error %d\n",
+			       drv->name, dev_name(dev), ret);
+		}
 	} else {
 		pr_debug("%s: probe of %s rejects match %d\n",
 		       drv->name, dev_name(dev), ret);
