diff --git a/sys_eeprom/24cXX.c b/sys_eeprom/24cXX.c
old mode 100644
new mode 100755
index be194ad..5bb5536
--- a/sys_eeprom/24cXX.c
+++ b/sys_eeprom/24cXX.c
@@ -44,7 +44,7 @@ static int i2c_write_2b(struct eeprom *e, __u8 buf[2])
 	r = i2c_smbus_write_byte_data(e->fd, buf[0], buf[1]);
 	if(r < 0)
 		fprintf(stderr, "Error i2c_write_2b: %s\n", strerror(errno));
-	usleep(1000);
+	usleep(5000);
 	return r;
 }
