diff --git a/include/configs/lenovo_g8296.h b/include/configs/lenovo_g8296.h
index cffdcbd..87136c0 100644
--- a/include/configs/lenovo_g8296.h
+++ b/include/configs/lenovo_g8296.h
@@ -657,6 +657,15 @@
 
 #include "configs/common_config.h"
 
+/*We run out of space so disable some of the options that ONIE enables*/
+#undef CONFIG_CMD_DHCP
+#undef CONFIG_CMD_EXT2
+#undef CONFIG_CMD_FAT
+#undef CONFIG_CMD_IMI
+#undef CONFIG_CMD_JFFS2
+#undef CONFIG_CMD_SETGETDCR
+#undef CONFIG_CMD_SOURCE
+
 /*
 ** Platform environment variables needed by the common env
 ** infrastructure.
