Don't use innetgr()

From: Abhisit Sangjan <abhisit.sangjan@gmail.com>


---
 access.c |    8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/access.c b/access.c
index b6afce3..7929d2c 100644
--- a/access.c
+++ b/access.c
@@ -37,10 +37,10 @@ static int match_hostname(const char **host_ptr, const char *addr, const char *t
 	if (!host || !*host)
 		return 0;
 
-#ifdef HAVE_INNETGR
-	if (*tok == '@' && tok[1])
-		return innetgr(tok + 1, host, NULL, NULL);
-#endif
+//#ifdef HAVE_INNETGR
+//	if (*tok == '@' && tok[1])
+//		return innetgr(tok + 1, host, NULL, NULL);
+//#endif
 
 	/* First check if the reverse-DNS-determined hostname matches. */
 	if (iwildmatch(tok, host))
