diff --git a/sys_eeprom/onie_tlvinfo.c b/sys_eeprom/onie_tlvinfo.c
old mode 100644
new mode 100755
index cf88f70..84cfc53
--- a/sys_eeprom/onie_tlvinfo.c
+++ b/sys_eeprom/onie_tlvinfo.c
@@ -733,7 +733,7 @@ int read_sys_eeprom(void *eeprom_data, int offset, int len)
 
     c = eeprom_data;
     if (eeprom_open(SYS_EEPROM_I2C_DEVICE, SYS_EEPROM_I2C_ADDR,
-		    EEPROM_TYPE_8BIT_ADDR, &e)) {
+		    EEPROM_TYPE_16BIT_ADDR, &e)) {
 	printf("ERROR: Cannot open I2C device\n");
 	return -1;
     }
@@ -759,7 +759,7 @@ int write_sys_eeprom(void *eeprom_data, int len)
     c = eeprom_data;
     for (i = 0; i < len; i++) {
 	if (eeprom_open(SYS_EEPROM_I2C_DEVICE, SYS_EEPROM_I2C_ADDR,
-			EEPROM_TYPE_8BIT_ADDR, &e)) {
+			EEPROM_TYPE_16BIT_ADDR, &e)) {
 	    printf("ERROR: Cannot open I2C device\n");
 	    return -1;
 	}
