Patch to add sys_eeprom information

diff --git a/sys_eeprom/sys_eeprom_platform.h b/sys_eeprom/sys_eeprom_platform.h
index 529f774..906257a 100644
--- a/sys_eeprom/sys_eeprom_platform.h
+++ b/sys_eeprom/sys_eeprom_platform.h
@@ -1,4 +1,10 @@
 #ifndef __SYS_EEPROM_PLATFORM_H_
 #define __SYS_EEPROM_PLATFORM_H_
 
+#define SYS_EEPROM_MAX_SIZE   256
+#define SYS_EEPROM_OFFSET     0
+#define SYS_EEPROM_SIZE       SYS_EEPROM_MAX_SIZE
+#define SYS_EEPROM_I2C_DEVICE "/dev/i2c-0"
+#define SYS_EEPROM_I2C_ADDR   0x51
+
 #endif /* __SYS_EEPROM_PLATFORM_H_ */
