Initail configurations for management port

diff --git a/drivers/net/ethernet/broadcom/tg3/tg3.c b/drivers/net/ethernet/broadcom/tg3/tg3.c
index 87a7911..4d06645 100644
--- a/drivers/net/ethernet/broadcom/tg3/tg3.c
+++ b/drivers/net/ethernet/broadcom/tg3/tg3.c
@@ -1628,6 +1628,7 @@ static void tg3_mdio_config_5785(struct tg3 *tp)
 static void tg3_mdio_start(struct tg3 *tp)
 {
 	tp->mi_mode &= ~MAC_MI_MODE_AUTO_POLL;
+	tp->mi_mode |= MAC_MI_MODE_SHORT_PREAMBLE;
 	tw32_f(MAC_MI_MODE, tp->mi_mode);
 	udelay(80);
 
@@ -2899,6 +2900,14 @@ static int tg3_phy_reset(struct tg3 *tp)
 		}
 	}
 
+	if (tp->phy_id == TG3_PHY_ID_BCM5718S) {
+		__tg3_writephy(tp, 0x8, 0x10, 0x1d0);
+		__tg3_writephy(tp, 0x1f, 0x00, 0x8140);
+		udelay(10000);
+		__tg3_writephy(tp, 0x1f, 0x1c, 0xfc0d);
+		__tg3_writephy(tp, 0x1f, 0x00, 0x1140);
+	}
+
 	if (tg3_flag(tp, 5717_PLUS) &&
 	    (tp->phy_flags & TG3_PHYFLG_MII_SERDES))
 		return 0;
